---
title: "Inspect Demultiplexing"
author: "Steve Pederson"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document: 
    fig_caption: yes
    fig_height: 6
    fig_width: 10
    toc: true
    toc_float: true
    code_folding: hide   
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE, message = FALSE
)
```

```{r loadPackages}
library(ngsReports)
library(magrittr)
library(scales)
library(pander)
library(tidyverse)
if (interactive()) setwd(here::here("R"))
```

```{r}
panderOptions("table.split.table", Inf)
panderOptions("big.mark", ",")
```


All plots from FastQC reports were generated in R using the Bioconductor package [ngsReports](https://bioconductor.org/packages/release/bioc/html/ngsReports.html)

```{r loadData}
rawFqc <- list.files("../0_rawData/FastQC/", pattern = "zip", full.names = TRUE) %>%
    FastqcDataList()
trimFqc <- list.files("../1_trimmed/FastQC/", pattern = "zip", full.names = TRUE) %>%
    FastqcDataList()
deMuxFqc <- list.files("../2_demux/FastQC/", pattern = "zip", full.names = TRUE) %>%
    FastqcDataList()
origFqc <- list.files("../previousAnalysis/FastQC/", pattern = "fastqc.zip", full.names = TRUE) %>%
    FastqcDataList()
samples <- read_tsv("../0_rawData/samples.tsv")
```

## Base Qualities

Each of the `r length(rawFqc[grepl("_1.fq.gz", fqName(rawFqc))])` paired-end, pooled libraries was inspected for overall quality.
Positions 3 & 4 from R2 libraries `FCC21WPACXX-CHKPEI13070002_L6` and `FCC21WPACXX-CHKPEI13070003_L7` showed clear problems with read qualities.
This was likely due to an unsatisfactory nucleotide diversity in the original sequencing run and not enough *phiX* to overcome this, particularly as all R2 reads will begin with the restriction site (i.e. fragments will terminate with this as there is no R2 barcode).

```{r plotBQRaw, fig.height=12, fig.cap = "*Base qualities before trimming*"}
plotBaseQuals(rawFqc, plotType = "boxplot")
```

Trimming was performed using `Adapter Removal v2.2.1` before de-multiplexing, with adapters automatically identified by `bbmerge` from `BBMap v6.62`.
Improvements were seen at the 3' end of the reads, but the previously noted issues at the 5' end were not specifically addressed during this step. 
During adapter removal, any reads <70nt were discarded as a complete pair, a quality trimming threshold of 20 was also used to discard low-quality reads.

```{r plotBQTrim, fig.height=12, fig.cap = "*Base qualities after trimming*"}
plotBaseQuals(trimFqc, plotType = "boxplot")
```


## Read Totals

```{r tableReadTotals}
list(
  readTotals(rawFqc) %>% mutate(Type = "Raw"),
  readTotals(trimFqc) %>% mutate(Type = "Trimmed")
) %>%
  bind_rows() %>%
  spread(key = "Type", value = "Total_Sequences") %>%
  mutate(Library = str_remove(Filename, "_[12].fq.gz")) %>%
  distinct(Library, .keep_all = TRUE) %>%
  dplyr::select(Library, Raw, Trimmed) %>%
  mutate(
    Retained = percent(Trimmed / Raw),
    Discarded = percent((Raw - Trimmed) / Raw)
  ) %>%
  arrange(Discarded) %>%
  bind_rows(
    tibble(
      Library = "Total",
      Raw = sum(.$Raw),
      Trimmed = sum(.$Trimmed)
    ) %>%
      mutate(
            Retained = percent(Trimmed / Raw),
            Discarded = percent((Raw - Trimmed) / Raw)
      )
  ) %>%
    pander(
      big.mark = ",",
      justify = "lrrrr",
      style = "rmarkdown",
      split.tables = Inf,
      emphasize.strong.rows = 8,
      caption = "*Results from adapter removal. Reads < 70bp after trimming were discarded during this process*"
    )
```

Demiltiplexing was performed using `sabre_pe` allowing for one mismatch in the barcode, and including the restriction site into the barcode sequence.
This demultiplixes based on the R1 reads only, and as such barcodes and restriction sites will be removed from all R1 reads.
No modification will occur for R2 reads, leading to differing read lengths for each sample (based on different barcode lengths), and for each read within the pair as R2 will be retained as full length.

The first check after demultiplexing is to ensure that read were not assigned to multiple individuals, given the permissive nature of this..
Read Totals before and after demultiplexing were then checked and the recovery rate was >93% for each library, with the clear exception of `FCC21WPACXX-CHKPEI13070002_L6`.
Manual inspection revealed that a significant majority (~9.3x10^6^ of 13.4x10^6^) of these non-recovered reads contained truncated barcodes with the first two bases missing.
If an additional recovery step was taken this would increase the recovery rate to 93% for this library, and may yield an additional 5-600,000 reads per sample.
For the next most poorly recovered library, an additional step may recover an additional 3x10^6^ reads. 
However, as this artefact was difficult to explain from the perspective of library preparation, no further action was taken at this point.

```{r readTotals}
trimReadTotals <- readTotals(trimFqc) %>%
    mutate(Library = str_remove_all(Filename, "_[12].fq.gz")) %>%
    distinct(Library, Total_Sequences)
deMuxReadTotals <- readTotals(deMuxFqc) %>%
    mutate(ID = str_remove_all(Filename, ".[12].fq.gz")) %>%
    distinct(ID, Total_Sequences) %>%
    left_join(samples, by = "ID") %>%
    group_by(Library) %>%
    summarise(DeMultiplexed = sum(Total_Sequences)) %>%
    filter(!is.na(Library))
```

```{r tableTrimmedReadTotals, echo = TRUE}
trimReadTotals %>%
  left_join(deMuxReadTotals) %>%
  mutate(`Recovery Rate` = DeMultiplexed / Total_Sequences) %>%
  dplyr::select(Library, everything()) %>%
  dplyr::rename(`Total Sequences` = Total_Sequences) %>%
  mutate(`Recovery Rate` = percent(`Recovery Rate`)) %>%
  arrange(`Recovery Rate`) %>%
  bind_rows(
    tibble(
      Library = "Total",
      `Total Sequences` = sum(.$`Total Sequences`),
      DeMultiplexed = sum(.$DeMultiplexed)
    ) %>%
      mutate(
        `Recovery Rate` = percent(DeMultiplexed / `Total Sequences`)
      ) 
  ) %>%
  pander(
    split.tables = Inf, 
    big.mark = ",",
    justify = "lrrr",
    style = "rmarkdown",
    emphasize.strong.rows = 8,
    caption = "*Recovery rate from demultiplexing the 1996 and 2012 samples, after adapter removal*"
  ) 
```

```{r cp}
cp <- paste("*Read Totals for each of the 1996 & 2012 samples.",
            "The black line indicates the mean library size across all libraries,",
            "whilst the dashed green line indicates the mean library size based",
            "of each individual library before demultiplexing.",
            "Bar colours indicate sample population as 1996 (blue) or 2012 (red).*")
```


```{r plotReadTotals, fig.cap=cp, fig.width=10}
plotly::ggplotly(
  deMuxFqc %>%
    magrittr::extract(grepl("(ora|gc).+1.fq.gz", fqName(.))) %>%
    readTotals() %>%
    mutate(ID = str_remove(Filename, ".1.fq.gz")) %>%
    left_join(samples) %>%
    mutate(Population = case_when(
      grepl("gc", ID) ~ "1996",
      grepl("ora", ID) ~ "2012"
    )) %>%
    ggplot(aes(ID, Total_Sequences, fill = Population)) +
    geom_bar(stat = "identity") +
    geom_hline(
      aes(yintercept = mn),
      data = . %>% summarise(mn = mean(Total_Sequences))
    ) +
    geom_hline(
      aes(yintercept = mn),
      data = . %>% group_by(Library) %>% summarise(mn = mean(Total_Sequences)),
      colour = "green",
      linetype = 2
    ) +
    facet_wrap(~Library, scales = "free_x", nrow = 2) +
    scale_y_continuous(labels = comma) +
    scale_fill_manual(values = c(rgb(0.1, 0.1, 0.7), rgb(0.7, 0.1, 0.1))) +
    labs(x = c(), y = c()) +
    theme_bw() +
    theme(
      axis.text.x = element_text(angle = 90),
      legend.position = "none"
    )
)
```

In summary, of the `r comma(sum(readTotals(rawFqc)$Total_Sequences)/2)` initial reads, a total of `r comma(sum(readTotals(deMuxFqc)$Total_Sequences)/2)` were retained after trimming and demultiplexing.
This represents a combined recovery rate of `r  percent(sum(readTotals(deMuxFqc)$Total_Sequences) / sum(readTotals(rawFqc)$Total_Sequences))` from the initial libraries.
The average number of reads recovered per sample was `r comma(mean(readTotals(deMuxFqc)$Total_Sequences))`, with all samples being between `r pander(comma(range(readTotals(deMuxFqc)$Total_Sequences)))` reads prior to alignment.

```{r}
deMuxFqc %>% 
  readTotals() %>% 
  mutate(ID = str_remove(Filename, ".[12].fq.gz")) %>% left_join(samples) %>% 
  distinct(ID, .keep_all = TRUE) %>% 
  filter(
    Total_Sequences == max(Total_Sequences) | Total_Sequences == min(Total_Sequences)
    ) %>%
  dplyr::select(ID, Barcode, `Total Sequences` = Total_Sequences, Library) %>%
  pander(
    justify = "llrl",
    caption = "The two samples with lowest and highest read recovery rates."
  )
```


### Comparison against previous analysis

This demultiplexing strategy was devised after previous unsatisfactory results were obtained using `process_radtags` from the `Stacks` pipeline.
The revised strategy showed considerable improvement in yield for the libraries `FCC21WPACXX-CHKPEI13070002_L6` and `FCC21WPACXX-CHKPEI13070003_L7` which both contained exculsively 1996 samples.
As such the revised strategy was chosen.

```{r plotLibComparison, fig.cap="*Comparison of library sizes after demultiplexing using sabre (salmon), vs the original method using process_radtags from the stacks pipeline (blue).*"}
list(
    origFqc %>%
        readTotals() %>%
        mutate(ID = str_remove(Filename, ".Lib..[12].fq"),
               Type = "stacks"),
    deMuxFqc %>%
        readTotals() %>%
        mutate(ID = str_remove(Filename, ".[12].fq.gz"),
               Type = "sabre")
) %>%
    bind_rows() %>%
    distinct(ID, Type, Total_Sequences) %>%
    dplyr::select(ID, Type, Total_Sequences) %>%
    filter(grepl("(gc|ora)", ID)) %>%
    mutate(Population = case_when(
            grepl("gc", ID) ~ "1996",
            grepl("ora", ID) ~ "2012"
        )) %>%
    left_join(samples) %>%
    ggplot(aes(Library, Total_Sequences, fill = Type)) +
    geom_boxplot() +
    labs(y = "Total Reads per Sample") +
    scale_y_continuous(labels = comma) +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```

```{r plotLibComp2, fig.cap="*Changes in library size using the improved recovery strategy. All samples and all libraries showed an improved rate of recovery.*"}
list(
    origFqc %>%
        readTotals() %>%
        mutate(ID = str_remove(Filename, ".Lib..[12].fq"),
               Type = "Previous"),
    deMuxFqc %>%
        readTotals() %>%
        mutate(ID = str_remove(Filename, ".[12].fq.gz"),
               Type = "Repeat")
) %>%
    bind_rows() %>%
    distinct(ID, Type, Total_Sequences) %>%
    dplyr::select(ID, Type, Total_Sequences) %>%
    filter(grepl("(gc|ora)", ID)) %>%
    spread("Type", "Total_Sequences") %>%
    mutate(Change = Repeat / Previous,
           Population = case_when(
            grepl("gc", ID) ~ "1996",
            grepl("ora", ID) ~ "2012"
        )) %>%
    left_join(samples) %>%
    ggplot(aes(ID, Change, fill = Population)) +
    geom_bar(stat = "identity") +
    geom_hline(yintercept = 1, linetype = 2) +
    facet_wrap(~Library, scales = "free_x", nrow = 2) +
    scale_y_continuous(labels = percent) +
    scale_fill_manual(values = c(rgb(0.1, 0.1, 0.7, 0.8), rgb(0.7, 0.1, 0.1, 0.8))) +
    labs(x = c(), y = "% Improvement") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90),
          legend.position = "none")
```

```{r plotLibComp3, fig.cap = "*Comparison of library sizes for the two populations.*"}
deMuxFqc %>%
  magrittr::extract(grepl("1.fq", fqName(.))) %>%
  readTotals()%>%
  mutate(Population = case_when(
    grepl("ora", Filename) ~ 2012,
    grepl("gc", Filename) ~ 1996
  ),
  Population = as.factor(Population)) %>%
  mutate(ID = str_remove(Filename, ".[12].fq.gz")) %>%
  ggplot(aes(Population, Total_Sequences, fill = Population)) +
  geom_boxplot() +
  scale_y_continuous(labels = comma) +
  labs(y = "Total Sequences") +
  theme_bw()
```


## GC Content


```{r plotGC, fig.cap="*GC content for all 1996 and 2012 samples.*"}
deMuxFqc %>%
    magrittr::extract(grepl("(ora|gc)", fqName(.))) %>%
    plotGcContent(plotType = "line", usePlotly = TRUE, theoreticalGC = FALSE)
```

Inspection of GC content showed that `gc2709` and `gc2700` appeared to have an exaggerated peak around 60%, whilst all other samples showed a more broad spread across the range.

From the Turretfield samples, collected for [a previous analysis](https://onlinelibrary.wiley.com/doi/abs/10.1111/mec.14228), `pt1125` showed an unexpected peak around 50% indicating that sample may contain reads from a different species.
This sample was recommended to be excluded from all further analysis.

```{r plotGcTF, fig.cap="*GC content for all Turretfield samples.*"}
origFqc %>%
    magrittr::extract(!grepl("(ora|gc)", fqName(.))) %>%
    plotGcContent(plotType = "line", usePlotly = TRUE, theoreticalGC = FALSE)
```


## Read Lengths

```{r plotLenDist, fig.cap = "*Length Distribution for all files*"}
deMuxFqc %>%
    plotSeqLengthDistn(usePlotly = TRUE, dendrogram = TRUE)
```


## Sequence Content

```{r plotSeqContent, cache = TRUE, fig.cap="*Sequence content showing the presence of the RS in both the Turretfield and R2 samples.*", fig.height=12}
plotSeqContent(deMuxFqc, usePlotly = TRUE, dendrogram = TRUE, cluster = TRUE)
```

## Conclusion

- The restriction site was subsequently removed from both R2 and Turretfield samples using `fastx_trimmer`, especially given the high error rate at positions 2/3 in some R2 samples
- No further processing is required as the presence of the restriction site at both ends means fixed length reads are *not required for Stacks*

## SessionInfo

```{r sessionInfo}
pander(sessionInfo())
```

