---
title: "Inspect Demultiplexing"
author: "Steve Pederson"
output: 
  html_document: 
    fig_caption: yes
    fig_height: 6
    fig_width: 10
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      warning = FALSE, message = FALSE)
```

## Setup

```{r loadPackages}
library(ngsReports)
library(magrittr)
library(scales)
library(pander)
library(tidyverse)
```

```{r loadData}
rawFqc <- list.files("../0_rawData/FastQC/", pattern = "zip", full.names = TRUE) %>%
    getFastqcData()
trimFqc <- list.files("../1_trimmed/FastQC/", pattern = "zip", full.names = TRUE) %>%
    getFastqcData()
deMuxFqc <- list.files("../2_demux/FastQC/", pattern = "zip", full.names = TRUE) %>%
    getFastqcData()
origFqc <- list.files("../previousAnalysis/FastQC/", pattern = "fastqc.zip", full.names = TRUE) %>%
    getFastqcData()
samples <- read_tsv("../0_rawData/samples.tsv")
```

## Base Qualities

Each of the libraries was inspected for overall quality.
Positions 3 & 4 from R2 libraries `FCC21WPACXX-CHKPEI13070002_L6` and `FCC21WPACXX-CHKPEI13070003_L7` showed clear problems with read qualities.
This was likely due to an unsatisfactory nucleotide diversity in the original sequencing run and not enough *phiX* to overcome this, particularly as all R2 reads will begin with the restriction site (i.e. fragments will terminate with this as there is no R2 barcode).

```{r, echo=FALSE, fig.height=12, fig.cap = "Base qualities before trimming"}
rawFqc %>% plotBaseQualities(plotType = "boxplot")
```

```{r, echo=FALSE, fig.height=12, fig.cap = "Base qualities after trimming"}
trimFqc %>% plotBaseQualities(plotType = "boxplot")
```


## Read Totals

```{r, echo=FALSE}
list(
    readTotals(rawFqc) %>% mutate(Type = "Raw"),
    readTotals(trimFqc) %>% mutate(Type = "Trimmed")
    ) %>%
    bind_rows() %>%
    spread(key = "Type", value = "Total_Sequences") %>%
    mutate(Library = str_remove(Filename, "_[12].fq.gz")) %>%
    distinct(Library, .keep_all = TRUE) %>%
    dplyr::select(Library, Raw, Trimmed) %>%
    mutate(Retained = percent(Trimmed / Raw),
           Discarded = percent((Raw - Trimmed) / Raw)) %>%
    arrange(Discarded) %>%
    pander(big.mark = ",",
           caption = "Results from adapter removal. Reads < 70bp after trimming were discarded during this process")
```


The first check after demultiplexing is to ensure that read were not assigned to multiple individuals by `sabre pe`, given that one mismatch was permitted per barcode/restriction-site.
Read Totals before and after demultiplexing were then checked and the recovery rate was >93% for each library, with the clear exception of `FCC21WPACXX-CHKPEI13070002_L6`.
Manual inspection revealed that a significant majority (~9.3x10^6^ of 13.4x10^6^) of these non-recovered reads contained truncated barcodes with the first two bases missing.
If an additional recovery step was taken this would increase the recovery rate to 93% for this library, and may yield an additional 5-600,000 reads per sample.
For the next most poorly recovered library, an additional step may recover an additional 3x10^6^ reads. 
However, no further action was taken at this point.

```{r readTotals}
trimReadTotals <- readTotals(trimFqc) %>%
    mutate(Library = str_remove_all(Filename, "_[12].fq.gz")) %>%
    distinct(Library, Total_Sequences)
deMuxReadTotals <- readTotals(deMuxFqc) %>%
    mutate(ID = str_remove_all(Filename, ".[12].fq.gz")) %>%
    distinct(ID, Total_Sequences) %>%
    left_join(samples, by = "ID") %>%
    group_by(Library) %>%
    summarise(DeMultiplexed = sum(Total_Sequences)) %>%
    filter(!is.na(Library))
```

```{r, echo = FALSE}
trimReadTotals %>%
    left_join(deMuxReadTotals) %>%
    mutate(`Recovery Rate` = DeMultiplexed / Total_Sequences) %>%
    dplyr::select(Library, everything()) %>%
    dplyr::rename(`Total Sequences` = Total_Sequences) %>%
    mutate(`Recovery Rate` = percent(`Recovery Rate`)) %>%
    arrange(`Recovery Rate`) %>%
    pander(split.tables = Inf, big.mark = ",",
           caption = "Recovery rate from demultiplexing the 1996 and 2012 samples") 
```

```{r, echo=FALSE}
cp <- paste("*Read Totals for each of the 1996 & 2012 samples.",
            "The black line indicates the mean library size,",
            "whilst the dashed green line indicates the mean library size based",
            "on the original library before demultiplexing.",
            "Bar colours indicate sample population as 1996 (blue) or 2012 (red).*")
```


```{r, echo=FALSE, fig.cap=cp}
plotly::ggplotly(
    deMuxFqc %>%
        magrittr::extract(grepl("(ora|gc).+1.fq.gz", fileName(.))) %>%
        readTotals() %>%
        mutate(ID = str_remove(Filename, ".1.fq.gz")) %>%
        left_join(samples) %>%
        mutate(Population = case_when(
            grepl("gc", ID) ~ "1996",
            grepl("ora", ID) ~ "2012"
        )) %>%
        ggplot(aes(ID, Total_Sequences, fill = Population)) +
        geom_bar(stat = "identity") +
        geom_hline(aes(yintercept = mn),
                   data = . %>% summarise(mn = mean(Total_Sequences))) +
        geom_hline(aes(yintercept = mn),
                   data = . %>% group_by(Library) %>% summarise(mn = mean(Total_Sequences)),
                   colour = "green",
                   linetype = 2) +
        facet_wrap(~Library, scales = "free_x", nrow = 2) +
        scale_y_continuous(labels = comma) +
        scale_fill_manual(values = c(rgb(0.1, 0.1, 0.7), rgb(0.7, 0.1, 0.1))) +
        labs(x = c(), y = c()) +
        theme_bw() +
        theme(axis.text.x = element_text(angle = 90),
              legend.position = "none")
)
```

### Comparison against previous analysis

The revised strategy showed considerable improvement in yield for the libraries `FCC21WPACXX-CHKPEI13070002_L6` and `FCC21WPACXX-CHKPEI13070003_L7` which both contained exculsively 1996 samples.
As such the revised strategy was pursued further.

```{r, echo=FALSE, fig.cap="Comparison of library sizes after demultiplexing using sabre (blue), vs the original method (salmon)."}
list(
    origFqc %>%
        readTotals() %>%
        mutate(ID = str_remove(Filename, ".Lib..[12].fq"),
               Type = "Previous"),
    deMuxFqc %>%
        readTotals() %>%
        mutate(ID = str_remove(Filename, ".[12].fq.gz"),
               Type = "Repeat")
) %>%
    bind_rows() %>%
    distinct(ID, Type, Total_Sequences) %>%
    dplyr::select(ID, Type, Total_Sequences) %>%
    filter(grepl("(gc|ora)", ID)) %>%
    mutate(Population = case_when(
            grepl("gc", ID) ~ "1996",
            grepl("ora", ID) ~ "2012"
        )) %>%
    left_join(samples) %>%
    ggplot(aes(Library, Total_Sequences, fill = Type)) +
    geom_boxplot() +
    labs(y = "Total Reads") +
    scale_y_continuous(labels = comma) +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```

```{r, echo=FALSE, fig.cap="Changes in library size using the improved recovery strategy"}
list(
    origFqc %>%
        readTotals() %>%
        mutate(ID = str_remove(Filename, ".Lib..[12].fq"),
               Type = "Previous"),
    deMuxFqc %>%
        readTotals() %>%
        mutate(ID = str_remove(Filename, ".[12].fq.gz"),
               Type = "Repeat")
) %>%
    bind_rows() %>%
    distinct(ID, Type, Total_Sequences) %>%
    dplyr::select(ID, Type, Total_Sequences) %>%
    filter(grepl("(gc|ora)", ID)) %>%
    spread("Type", "Total_Sequences") %>%
    mutate(Change = Repeat / Previous,
           Population = case_when(
            grepl("gc", ID) ~ "1996",
            grepl("ora", ID) ~ "2012"
        )) %>%
    left_join(samples) %>%
    ggplot(aes(ID, Change, fill = Population)) +
    geom_bar(stat = "identity") +
    geom_hline(yintercept = 1, linetype = 2) +
    facet_wrap(~Library, scales = "free_x", nrow = 2) +
    scale_y_continuous(labels = percent) +
    scale_fill_manual(values = c(rgb(0.1, 0.1, 0.7, 0.8), rgb(0.7, 0.1, 0.1, 0.8))) +
    labs(x = c(), y = "% Improvement") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90),
          legend.position = "none")
```

```{r, echo=FALSE, fig.cap = "Comparison of library sizes for the three populations."}
list(
    origFqc %>%
        magrittr::extract(!grepl("(ora|gc)", fileName(.))) %>%
        readTotals(),
    deMuxFqc %>%
        magrittr::extract(grepl("1.fq", fileName(.))) %>%
        readTotals()
) %>%
    bind_rows() %>%
    mutate(Population = case_when(
        grepl("ora", Filename) ~ 2012,
        grepl("gc", Filename) ~ 1996,
        !grepl("(ora|gc)", Filename) ~ 2010
    ),
    Population = as.factor(Population)) %>%
    mutate(ID = str_remove(Filename, ".[12].fq.gz")) %>%
    ggplot(aes(Population, Total_Sequences, fill = Population)) +
    geom_boxplot() +
    scale_y_continuous(labels = comma) +
    theme_bw()
```


## GC Content


```{r, echo=FALSE, fig.cap="GC content for all 1996 and 2012 samples."}
deMuxFqc %>%
    magrittr::extract(grepl("(ora|gc)", fileName(.))) %>%
    plotGcContent(plotType = "line", usePlotly = TRUE, theoreticalGC = FALSE)
```

Inspection by GC content showed that `gc2709` and `gc2700` appeared to have an exaggerated peak around 60%, whilst all other samples showed a more broad spread across the range.
From the Turretfield samples (previous analysis), `pt1125` showed an unexpected peak around 50% indicating that sample may contain reads from a different species.
This sample should be excluded from all further analysis.

```{r, echo=FALSE, fig.cap="GC content for all Turretfield samples."}
origFqc %>%
    magrittr::extract(!grepl("(ora|gc)", fileName(.))) %>%
    plotGcContent(plotType = "line", usePlotly = TRUE, theoreticalGC = FALSE)
```


## Read Lengths

```{r, echo=FALSE, fig.cap = "Length Distribution for all files"}
deMuxFqc %>%
    plotSequenceLengthDistribution(usePlotly = TRUE, dendrogram = TRUE)
```


## Sequence Content

```{r, echo=FALSE, fig.cap="Sequence content showing the presence of the RS in both the Turretfield and R2 samples.", fig.height=12}
plotSequenceContent(deMuxFqc, usePlotly = TRUE)
```

## Conclusion

- The restriction site should be removed from both R2 and Turretfield samples, especially given the high error rate at positions 2/3 in some R2 samples
- No further processing is required as the presence of the restriction site at both ends means fixed length reads are not required for Stacks