---
title: "Analysis of SNPs"
author: "Steve Pederson"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document: 
    fig_caption: yes
    fig_height: 8
    fig_width: 10
    toc: true
    toc_float: true
    code_folding: hide 
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE, message = FALSE
)
```


```{r loadPackages}
library(SeqArray)
library(SNPRelate)
library(pander)
library(scales)
library(magrittr)
library(tidyverse)
library(readxl)
library(sp)
library(ggmap)
library(rgdal)
library(ggsn)
library(parallel)
library(qqman)
library(ggrepel)
library(plyranges)
library(rtracklayer)
library(AnnotationDbi)
library(GO.db)
```


```{r setOptions}
if (interactive()) here::here("R") %>% setwd()
theme_set(theme_bw())
panderOptions("missing", "")
mc <- min(12, detectCores() - 1)
alpha <- 0.05
maxKb <- 40
```

```{r tidyP}
tidyP <- function(p, m = 0.001){
  x <- rep("", length(p))
  x[p < m] <- sprintf("%.2e", p[p < m])
  x[p >= m] <- sprintf("%.3f", p[p >= m])
  x
}
```

```{r keepSNPs}
keepSNPs <- readRDS("keepSNPsAfterLDPruning.RDS")
```


```{r pdfParams}
w <- 169/25.4
h <- 169/25.4
```

This workflow immediately follows `03_SNPFiltering` and performs an analysis on the `r nrow(keepSNPs)` SNPs retained after filtering.
Analysis performed were:

- To assess the rate of missing values by retained SNP and sample
- Run a PCA analysis using SNPs & samples 
- Analyse using comparisons between genotypes and allele frequencies, as well as using Bayescan and FLK

```{r loadGDS}
gdsPath <- file.path("..", "5_stacks", "gds", "populations.snps.gds")
gdsFile <- seqOpen(gdsPath, readonly = TRUE)
```

```{r loadFilteredSNPs}
sampleID <- tibble(
    Sample = seqGetData(gdsFile, "sample.annotation/Sample"),
    Population = seqGetData(gdsFile, "sample.annotation/Population"),
    Location = seqGetData(gdsFile, "sample.annotation/Location")
)
popSizes <- sampleID %>% 
    group_by(Population) %>%
    summarise(n = dplyr::n())
```


```{r genotypes, include=FALSE}
snps <- tibble(
    variant.id = seqGetData(gdsFile, "variant.id") ,
    chromosome = seqGetData(gdsFile, "chromosome"),
    position = seqGetData(gdsFile, "position"),
    snpID = seqGetData(gdsFile, "annotation/id")
) %>%
    mutate(
      snpID = str_remove(snpID, ":.$"),
      snpID = str_replace(snpID, ":", "_")
    ) %>%
    separate(snpID, into = c("Locus ID", "Col"), remove = FALSE) %>%
    mutate_at(vars(`Locus ID`, Col), as.integer) %>%
    mutate(
      Col = Col - 1,
      snpID = paste(`Locus ID`, Col, sep = "_")
    ) %>%
  right_join(keepSNPs)
seqSetFilter(gdsFile, variant.id = snps$variant.id)
genotypes <- snps %>%
    cbind(
      seqGetData(gdsFile, "genotype") %>% 
        apply(MARGIN = 3, colSums) %>%
        t %>%
        set_colnames(sampleID$Sample)) %>%
  as_tibble() %>%
  gather(Sample, Genotype, one_of(sampleID$Sample)) %>%
  dplyr::filter(!is.na(Genotype)) %>%
  arrange(variant.id, Sample) %>%
  left_join(sampleID)
seqResetFilter(gdsFile)
```

- After loading the data from the previous section, an additional filtering step was performed, where SNPs were required to have a non-zero MAF in the initial 1996 population.

```{r snpIn1996}
snpIn1996 <- genotypes %>% 
    filter(Population == 1996) %>% 
    group_by(variant.id) %>% 
    summarise(maf = mean(Genotype) / 2) %>%
    filter(maf > 0)
genotypes %<>% 
    right_join(snpIn1996)
```

- This additional filtering step restricted analysis to `r comma(nrow(snpIn1996))` SNPs instead of the initial `r comma(nrow(keepSNPs))` SNPs.

```{r closeGDS}
seqClose(gdsFile)
```

In addition to the above, the set of genes corresponding the Ensembl Release 96 were loaded.

```{r ensGenes}
ensGenes <- file.path(
  "..", "external", "Oryctolagus_cuniculus.OryCun2.0.96.gff3.gz"
) %>%
  import.gff3(feature.type = "gene", sequenceRegionsAsSeqinfo = TRUE) %>%
  .[,c("gene_id", "Name", "description")]
```



```{r snpsGR}
snpsGR <- makeGRangesFromDataFrame(
    df = snps, 
    keep.extra.columns = TRUE, 
    ignore.strand = TRUE, 
    seqinfo = seqinfo(ensGenes), 
    seqnames.field = "chromosome", 
    start.field = "position", 
    end.field = "position"
)
```

The set of SNPs under investigation was also defined as a GRanges object.
This was intersected with the set of genes to connect SNPs to genes within `r maxKb` kb.

```{r snp2Gene}
snp2Gene <- suppressWarnings(
  snpsGR %>% 
    resize(width = 2*1000*maxKb - 1, fix = "center") %>%
    trim() %>%
    join_overlap_inner(ensGenes) 
)
```

A set of SNPs within exonic regions was also defined in order to detect any SNPs within coding regions.

```{r ensExons}
ensExons <- file.path(
  "..", "external", "Oryctolagus_cuniculus.OryCun2.0.96.gff3.gz"
) %>%
  import.gff3(feature.type = "exon", sequenceRegionsAsSeqinfo = TRUE) %>%
  join_overlap_inner_within(
    ensGenes, maxgap = 0, minoverlap = 0, suffix = c(".exon", "")
  ) %>%
  join_overlap_inner(snpsGR)
ensExons <- ensExons[,setdiff(colnames(mcols(ensExons)), "Name.exon")]
```


## PCA

```{r lowCall}
lowCall <- c("gc2901", "gc2776", "gc2731", "gc2727", "gc2686")
```


```{r pca, cache = TRUE}
snp4PCA <- genotypes %>% 
    filter(!Sample %in% lowCall) %>%
    group_by(variant.id, Population) %>% 
    summarise(n = dplyr::n()) %>% 
    spread(Population, n) %>%
    mutate(N = `1996` + `2010` + `2012`) %>% 
    ungroup() %>%
    filter(N > 0.95*(sum(popSizes$n) - length(lowCall)))
pca <- genotypes %>%
    filter(variant.id %in% snp4PCA$variant.id) %>%
    dplyr::select(variant.id, Sample, Genotype) %>%
    spread(Sample, Genotype) %>%
    as.data.frame() %>%
    column_to_rownames("variant.id") %>%
    as.matrix() %>%
    apply(2, function(x){
        x[is.na(x)] <- mean(x, na.rm = TRUE)
        x
    }) %>%
    t() %>%
    .[, apply(., 2, function(x){length(unique(x)) > 1})] %>%
    prcomp( center = TRUE)
```

As noted in the previous section sample samples `r pander(lowCall)` had a SNP identification rate $< 50$% and as such these were marked as potential outliers.
Ignoring these samples, and restricting data to SNPs identified in $>95$% of all samples, a preliminary PCA was performed
This amounted to `r comma(nrow(snp4PCA))` of the possible `r comma(length(unique(genotypes$variant.id)))` SNPs for analysis using PCA.
Missing values were specified as the mean MAF over all populations combined.

Given the initially observed structure, in which samples from the 2012 population are separating from the other samples which group with the 1996 population, the collection points for the 2012 samples as investigated.

```{r addLatLong}
sampleID %<>%
    left_join(
        file.path("..", "external", "GPS_Locations.xlsx") %>%
            read_excel() %>%
            dplyr::select(Sample, ends_with("tude")) %>%
            mutate(Sample = gsub("[Oo][Rr][Aa] ([0-9ABC]*)", "ora\\1", Sample))
    )
```


```{r plotPCA, cache=TRUE, dependson='pca', fig.cap="*PCA showing population structure. Point size reflects the proportion of SNPs for which imputation was required, and the observed structure appeared independent of this.*"}
fs <- 14
ps <- 3
popCols <- c(rgb(0, 0, 0, 0.7), 
             rgb(0.8, 0.2, 0.2, 0.9), 
             rgb(0.2, 0.8, 0.2, 0.9),
             rgb(0.5, 0.5, 1, 0.7))
pca4Plot <- pca$x %>%
    as.data.frame() %>%
    rownames_to_column("Sample") %>%
    as_tibble() %>%
    dplyr::select(Sample, PC1, PC2, PC3) %>%
    left_join(sampleID) %>%
    mutate(Cluster = kmeans(cbind(PC1, PC2, PC3), 3)$cluster) %>%
    group_by(Cluster) %>%
    mutate(maxY = max(Latitude, na.rm = TRUE)) %>%
    ungroup() %>%
    mutate(
      Population = case_when(
        Population == 1996 ~ "1996",
        Population == 2010 ~ "Outgroup (Turretfield)",
        maxY == max(maxY) ~ "2012 (Outer)",
        maxY != max(maxY) ~ "2012 (Central)"
      )
    ) %>%
  left_join(
    genotypes %>% 
      filter(variant.id %in% snp4PCA$variant.id, !Sample %in% lowCall) %>% 
      group_by(Sample) %>% 
      tally() %>% 
      mutate(imputationRate = 1 - n / nrow(snp4PCA)
      )
  ) 
pca4Plot %>%
    ggplot(aes(PC1, PC2, colour = Population, size = imputationRate)) +
    geom_point(alpha = 0.6) +
    labs(
      x = paste0("PC1 (", percent(summary(pca)$importance[2,"PC1"]), ")"),
      y = paste0("PC2 (", percent(summary(pca)$importance[2,"PC2"]), ")"),
      size = "Imputation Rate"
    ) +
  scale_colour_manual(values = popCols)
```

```{r mapParams}
loc <- c(
  range(sampleID$Longitude, na.rm = TRUE) %>% mean,
  range(sampleID$Latitude, na.rm = TRUE) %>% mean
)
saPoly <- readRDS(file.path("..", "external", "saPoly.RDS"))
roads <- readRDS(file.path("..", "external", "roads.RDS"))
gc <- SpatialPoints(cbind(x = 138.655972, y = -31.200305))
proj4string(gc) <- "+proj=longlat +ellps=GRS80 +no_defs"
xBreaks <- seq(138.65, 138.8, by = 0.05)
xLabs <- parse(text = paste(xBreaks, "*degree ~ E", sep = ""))
yBreaks <- seq(-31.2, -31.32, by = -0.04)
yLabs <- parse(text = paste(-yBreaks, "*degree ~ S", sep = ""))
leftN <- tibble(
  x = c(138.7965, 138.8, 138.8) - 0.01,
  y = c(-31.2, -31.198, -31.193)
)
rightN <- tibble(
  x = c(138.8, 138.8, 138.8035) - 0.01,
  y = c(-31.193, -31.198, -31.2)
)
```

```{r mainPlot, cache=TRUE}
ggMap <- get_map(loc, zoom = 12, maptype = "terrain", color = "bw")
zoomPlot <- ggmap(ggMap, extent = "normal", maprange = FALSE) +
    geom_path(
        aes(long, lat, group = group), 
        data = subset(roads, SURFACE == "UNSE"), 
        linetype = 2, size = 0.3) + 
    geom_path(
        aes(long, lat, group = group), 
        data = subset(roads, SURFACE != "UNSE"), 
        linetype = 1, size = 0.4) +
    geom_label(x = 138.74, y = -31.29, label = "Flinders Ranges NP", alpha = 0.4) +
    geom_label(x = 138.72, y = -31.22, label = "Gum Creek", alpha = 0.4) +
    geom_point(aes(x, y), data = as.data.frame(gc), shape = 3, size = 3) +
    geom_text(aes(x, y), label = "HS", data = as.data.frame(gc), nudge_y = 0.005) +
    geom_polygon(
        data = subset(saPoly, F_CODE == "HD"),
        aes(long, lat, group = group),
        fill = rgb(1, 1, 1, 0), colour = "grey10", size = 0.3) +
    geom_polygon(
        data = subset(saPoly, F_CODE == "PARK"),
        aes(long, lat, group = group),
        fill = rgb(1, 1, 1, 0), colour = "grey10", size = 0.2) +
  geom_point(
    data= filter(pca4Plot, grepl("2012", Population)),
    aes(Longitude, Latitude, colour = Population),
    size = 0.9*ps) +
  geom_polygon(
    aes(x, y), data = leftN, fill = "white", colour = "grey10", size = 0.4
  ) +
  geom_polygon(
    aes(x, y), data = rightN, fill = "grey10", colour = "grey10", size = 0.4
  ) +
  geom_text(
    x = 138.79, y = -31.19, label = "N", colour = "grey10", size = 4
  ) +
  scale_colour_manual(values = popCols[2:3]) +
  coord_cartesian(
    xlim = c(138.618, 138.81),
    ylim = c(-31.335, -31.18),
    expand = 0
  ) +
  scale_x_continuous(breaks = xBreaks, labels = xLabs) +
  scale_y_continuous(breaks = yBreaks, labels = yLabs) +
  guides(colour = FALSE) +
  ggsn::scalebar(
    x.min = 138.618,
    x.max = 138.81,
    y.min = -31.335, 
    y.max = -31.18,
    transform = TRUE,
    dist = 2, 
    dist_unit = "km",
    model = 'GRS80',
    height = 0.012, 
    st.size = 4,
    location = 'bottomright',
    anchor = c(x = 138.8, y = -31.328)) +
  labs(x = "Longitude", y = "Latitude") +
  theme(
    text = element_text(size = fs),
    plot.margin = unit(c(1, 1, 1, 1), "mm")
  )
```


```{r Map4Inset, cache=TRUE, dependson='mainPlot'}
ausPolygon <- readRDS(file.path("..", "external", "ausPolygon.RDS"))
ausPts <- SpatialPoints(cbind(x = loc[1], y = loc[2]))
proj4string(ausPts) <- proj4string(ausPolygon)
ausPlot <- ggplot() +
  geom_polygon(
    data = ausPolygon, 
    aes(long, lat, group = group),
    fill = "white", colour = "black"
  ) + 
  geom_point(data = as.data.frame(ausPts), aes(x, y), size = 1.5) +
  theme_void() +
  theme(plot.background = element_rect(fill = "white", colour = "black"))
```

```{r map2Pdf, cache=TRUE, dependson=c('mainPlot', 'Map4Inset'), include=FALSE}
pdf(file.path("..", "figures", "Figure1.pdf"), width = w, height = h) 
grid.newpage()
vp1 <- viewport(0.5,0.5, 1, 1)
vp2 <- viewport(0.32, 0.25, 0.33, 0.33)
print(zoomPlot, vp = vp1)
print(ausPlot, vp = vp2)
dev.off()
```

```{bash}
convert \
  -density 600 \
  -size 1920 \
  ../figures/Figure1.pdf \
  ../figures/Figure1.png
```


```{r plotWithInset_png, fig.cap="*Figure 1: Collection points for all 2012 samples with colours showing sub-populations initially defined by PCA analysis and *k*-means clustering.*", out.width='80%'}
knitr::include_graphics(file.path("..", "figures", "Figure1.png"))
```

```{r zoomCentralPlot, cache=TRUE, fig.height=7, fig.width=7, fig.cap = "*Zoomed-in view of the central region for 2012 samples with colours showing sub-populations defined by PCA analysis. The region considered to be the Central Region is shaded in red. Due to overlapping GPS points a small amount of jitter has been added to the x-axis.*"}
zoomLoc <- c(138.753, -31.242)
xBreaks <- seq(138.74, 138.76, by = 0.01)
xLabs <- parse(text = paste(xBreaks, "*degree ~ W", sep = ""))
yBreaks <- seq(-31.235, -31.25, by = -0.005)
yLabs <- parse(text = paste(-yBreaks, "*degree ~ S", sep = ""))
central <- rbind(
  x = c(138.749, 138.755),
  y = c(-31.2365, -31.2495)
) %>%
  set_colnames(c("min", "max"))
leftN <- tibble(
  x = c(138.7595, 138.76, 138.76),
  y = c(-31.234, -31.2335, -31.2325)
)
rightN <- tibble(
  x = c(138.76, 138.76, 138.7605),
  y = c(-31.2325, -31.2335, -31.234)
)
get_map(zoomLoc, zoom = 15, maptype = "terrain", source = "google", color = "bw") %>%
  ggmap() +
  geom_jitter(
    data = filter(pca4Plot, grepl("2012", Population)), 
    aes(Longitude, Latitude, colour = Population), 
    size = 3, width = 0.0005, height = 0) +
  geom_rect(
    xmin = central["x", "min"],
    xmax = central["x", "max"],
    ymin = central["y", "min"],
    ymax = central["y", "max"],
    fill = "red", alpha = 0.01, colour = "black") +
  geom_polygon(
    aes(x, y), data = leftN, fill = "white", colour = "grey10", size = 0.4
  ) +
  geom_polygon(
    aes(x, y), data = rightN, fill = "grey10", colour = "grey10", size = 0.4
  ) +
  geom_text(
    x = 138.76, y = -31.232, label = "N", colour = "grey10", size = 5
  ) +
  scale_colour_manual(values = popCols[2:3]) +
  scale_x_continuous(breaks = xBreaks, labels = xLabs) +
  scale_y_continuous(breaks = yBreaks, labels = yLabs) +
  theme_bw() +
  guides(colour = FALSE) +
  labs(x = "Longitude", y = "Latitude") +
  coord_cartesian(
    xlim = c(138.74, 138.762),
    ylim = c(-31.253, -31.231),
    expand = 0
  ) +
  ggsn::scalebar(
    x.min = 138.74, x.max = 138.762, 
    y.min = -31.253, y.max = -31.231,
    transform = TRUE,
    dist = 0.25, dist_unit = "km",, model = 'GRS80', 
    height = 0.012, st.size = 4,
    location = 'bottomright',
    anchor = c(x = 138.761, y = -31.252)
    )
```

## Region Analysis

### Removal of SNPs Associated with Collection Region

The structure observed within the 2012 population in the PCA could possibly be explained by recent migration into this region.
As the samples collected in the outer regions appeared very similar to the 1996 population in the above plots, this would possibly indicate migration is a very recent event as the genetic influence of this has not spread through the wider area.
Although this may be due to other factors such as sampling bias, this structure was addressed by identifying SNPs which showed an association with the sub-populations identified by PCA analysis.
In this way, any candidate SNPs obtained below will be less impacted by this structure, and will be more reflective of the intended variable under study, i.e. selection over time, as opposed to any internal structure of the 2012 population which may indicate migration into the breeding population.

```{r oraRegions}
oraRegions <- pca4Plot %>% 
  filter(grepl("2012", Population)) %>% 
  rowwise() %>%
  mutate(
    yCentral = cut(Latitude, breaks = central["y",], include.lowest = TRUE), 
    xCentral = cut(Longitude, breaks = central["x",], include.lowest = TRUE),
    Central = (is.na(yCentral) + is.na(xCentral)) == 0
  ) %>%
  dplyr::select(Sample, Central) 
```


#### Testing for Structure in 2012

This model tests:  
H<sub>0</sub>: No association between genotypes and collection region  
H<sub>A</sub>: An association exists between genotypes and collection region

```{r regionResults}
regionResults <- genotypes %>%
    filter(Population == 2012) %>%
    split(f = .$variant.id) %>%
    mclapply(
      function(x){
        ft <- list(p.value = NA)
        if (length(unique(x$Genotype)) > 1) {
          ft <- x %>%
            left_join(oraRegions) %>%
            group_by(Genotype, Central) %>%
            tally() %>%
            spread(Genotype, n, fill = 0) %>%
            column_to_rownames("Central") %>%
            fisher.test()
        }
        x %>%
          distinct(variant.id, snpID, chromosome, position) %>%
          mutate(p = ft$p.value)
      }, 
      mc.cores = mc
    ) %>%
  bind_rows() %>%
  filter(!is.na(p)) %>%
  arrange(p)
```


A total of `r sum(regionResults$p < 0.05)` SNPs were detected as showing a significant association between genotype and the collection region.
Under H<sub>0</sub>, the number expected using &#945; = 0.05 would be `r floor(0.05*length(unique(genotypes$variant.id)))`, and as this number was approximately double that expected, this was taken as evidence of this being a genuine point of concern for this dataset.

Notably, Type II errors were of principle concern in this instance, and as such every SNP with p < 0.05 in the above test was excluded from downstream analysis.

```{r regionSNPs}
regionSNPs <- filter(regionResults, p < 0.05)
saveRDS(regionSNPs, "regionSNPs.RDS")
```

Under this additional filtering step, **the original set of `r nrow(regionResults)` SNPs will be reduced to `r comma(sum(regionResults$p >= 0.05))`** for testing by genotype and allele frequency.

#### Verification Of Removal

In order to verify that the removal of the above SNPs removed the undesired population structure from the 2012 population, the above PCA was repeated, excluding the SNPs marked for removal.
The previous structure noted in the data was no longer evident, and as such, these SNPs were marked for removal during analysis by genotype and allele frequency. 


```{r pcaPost}
pcaPost <- genotypes %>%
  filter(
    variant.id %in% snp4PCA$variant.id,
    !variant.id %in% regionSNPs$variant.id
  ) %>%
  dplyr::select(variant.id, Sample, Genotype) %>%
  spread(Sample, Genotype) %>%
  as.data.frame() %>%
  column_to_rownames("variant.id") %>%
  as.matrix() %>%
  apply(2, function(x){
    x[is.na(x)] <- mean(x, na.rm = TRUE)
    x
  }) %>%
  t() %>%
  .[, apply(., 2, function(x){length(unique(x)) > 1})] %>%
  prcomp( center = TRUE) 
pcaPost4Plot <- pcaPost$x %>%
  as.data.frame() %>%
  rownames_to_column("Sample") %>%
  as_tibble() %>%
  dplyr::select(Sample, PC1, PC2, PC3) %>%
  left_join(sampleID) %>%
  mutate(Cluster = kmeans(cbind(PC1, PC2, PC3), 3)$cluster) %>%
  group_by(Cluster) %>%
  mutate(maxY = max(Latitude, na.rm = TRUE)) %>%
  ungroup() %>%
  mutate(
    Population = case_when(
      Population == 1996 ~ "1996",
      Population == 2010 ~ "Outgroup (Turretfield)",
      maxY == max(maxY) ~ "2012 (Outer)",
      maxY != max(maxY) ~ "2012 (Central)"
    )
  ) %>%
  left_join(
    genotypes %>% 
      filter(variant.id %in% snp4PCA$variant.id, !Sample %in% lowCall) %>% 
      group_by(Sample) %>% 
      tally() %>% 
      mutate(imputationRate = 1 - n / nrow(snp4PCA))
  ) 
```


```{r comparePCA, results='hide', include=FALSE}
pdf(file.path("..", "figures", "Figure2.pdf"), height = h, width = w*1.8)
bind_rows(
  pca4Plot %>%
    dplyr::select(Sample, PC1, PC2, Population) %>%
    mutate(Status = "Pre"),
  pcaPost4Plot %>%
    dplyr::select(Sample, PC1, PC2, Population) %>%
    mutate(Status = "Post")
) %>%
  left_join(oraRegions) %>%
  mutate(Status = factor(Status, levels = c("Pre", "Post"))) %>%
  ggplot(aes(PC1, PC2, colour = Population)) +
  geom_point() +
  facet_wrap(~Status) +
  scale_colour_manual(values = popCols) +
  theme(
    legend.position = "top",
    text = element_text(size = fs*1.5),
    strip.text = element_text(size = fs*1.4),
    axis.title = element_text(size = fs*1.4)
  )
dev.off()
```

```{bash}
convert \
-density 600 \
-size 1920 \
../figures/Figure2.pdf \
../figures/Figure2.png
```


```{r Figure2_png, fig.cap="*Figure 2: Principal Components Analysis showing structures before removal of SNPs denoting collection region in the 2012 population, and after removal of these SNPs*", out.width='80%'}
knitr::include_graphics(file.path("..", "figures", "Figure2.png"))
```

## SNP Analysis

### Genotype Frequency Model

This model tests:  
H<sub>0</sub>: No association between genotypes and populations  
H<sub>A</sub>: An association exists between genotypes and populations

```{r genotypeResults}
genotypeResults <- genotypes %>%
  filter(
    Population != 2010,
    !variant.id %in% regionSNPs$variant.id
  ) %>%
  group_by(variant.id, snpID, Population, Genotype) %>%
  tally() %>%
  ungroup() %>%
  split(f = .$variant.id) %>%
  mclapply(function(x){
    ft <- list(p.value = NA)
    if (length(unique(x$Genotype)) > 1) {
      ft <- x %>%
        spread(Genotype, n, fill = 0) %>%
        column_to_rownames("Population") %>%
        dplyr::select(-variant.id, -snpID) %>%
        fisher.test()
    }
    x %>% 
      distinct(variant.id) %>%
      mutate(p = ft$p.value)
  },
  mc.cores = mc
  ) %>%
  bind_rows() %>%
  filter(!is.na(p)) %>%
  mutate(FDR = p.adjust(p, "fdr"), adjP = p.adjust(p, "bonferroni")) %>%
  arrange(p) %>%
  left_join(
    genotypes %>%
      distinct(variant.id, snpID, chromosome, position)
  ) %>%
  dplyr::select(variant.id, snpID, chromosome, position, everything())
```

```{r genoSNPs}
genoSNPs <- genotypeResults %>%
  dplyr::filter(FDR < alpha) %>%
  .[["snpID"]]
```


```{r}
snpsGR %>% 
  subset(snpID %in% genoSNPs) %>% 
  join_nearest(ensGenes) %>%
  as.data.frame() %>%
  left_join(
    ensGenes %>% as.data.frame(),
    by = c("Name", "gene_id", "seqnames", "description"),
    suffix = c("_snp", "_gene")
  ) %>%
  as_tibble() %>%
  dplyr::select(seqnames, start_snp, variant.id, snpID, gene_id, Name, start_gene, end_gene) %>%
  mutate(dist2Gene = case_when(
    start_snp >= start_gene & start_snp <= end_gene ~ 0L,
    start_snp < start_gene ~ start_gene - start_snp,
    start_snp > end_gene ~ start_snp - end_gene
  ),
  Name = case_when(
    dist2Gene > maxKb*1000 ~ "",
    dist2Gene < maxKb*1000 ~ Name
  ),
    gene_id = case_when(
    dist2Gene > maxKb*1000 ~ "",
    dist2Gene < maxKb*1000 ~ gene_id
  ),
  inExon = snpID %in% ensExons$snpID,
  dist2Gene = case_when(
    Name == "" ~ "",
    inExon ~ "CDS",
    dist2Gene == 0 & !inExon ~ "Intronic",
    dist2Gene > 0 ~ comma(dist2Gene)
  ),
  seqnames = as.character(seqnames)
  ) %>%
  dplyr::select(
    snpID,
    chromosome = seqnames,
    position = start_snp,
    gene_id, Name, dist2Gene
  ) %>%
  bind_rows(
    subset(snpsGR, snpID %in% setdiff(genoSNPs, .$snpID)) %>%
      as.data.frame() %>%
      mutate(seqnames = as.character(seqnames)) %>%
      dplyr::select(
        snpID,
        chromosome = seqnames,
        position = start
      )
  ) %>%
  left_join(genotypeResults) %>%
  dplyr::select(-variant.id, FDR) %>%
  arrange(p) %>%
  mutate(
    position = comma(position),
    p = tidyP(p),
    FDR = tidyP(FDR),
    adjP = tidyP(adjP)
    ) %>%
  pander(
    justify = "llrlllrrr",
    style = "rmarkdown",
    big.mark = ",",
    split.tables = Inf,
    caption = paste(
      "*Candidate SNPs to an FDR of", percent_format(1)(alpha), "when analysing by genotype.",
      "The nearest gene to each SNP (within", paste0(maxKb, "kb"), ") is indicated.",
      "Bonferroni-adjusted p-values are also given in the final column.*")
  )
```


Under the full genotype model:

- `r nrow(filter(genotypeResults, adjP < alpha))` genotypes were detected as being significantly associated with the two populations when controlling the FWER at &#945; = `r alpha` (using the Bonferroni adjustment)
- `r nrow(filter(genotypeResults, FDR < alpha))` genotypes were detected as being significantly associated with the two populations when controlling the FDR at &#945; = `r alpha`
- For the most highly ranked SNP (`r genotypeResults$snpID[[1]]`), the minor allele has been completely lost in the 2012 population


```{r manhattanGenotype, results='hide', include=FALSE}
pdf(file.path("..", "figures", "FigureS3a.pdf"), height = 0.6*h, width = w)
hLine <- max(filter(genotypeResults, FDR < alpha)$p)
hLight <- filter(genotypeResults, adjP < alpha, chromosome %in% 1:21)$variant.id
genotypeResults %>%
  mutate(Chr = as.numeric(chromosome)) %>%
  filter(!is.na(Chr)) %>%
  dplyr::select(SNP = variant.id, CHR = Chr, BP = position, P = p) %>%
  manhattan(suggestiveline = FALSE, 
            genomewideline = -log10(hLine),
            highlight = hLight)
dev.off()
```

```{bash}
convert \
-density 600 \
-size 1920 \
../figures/FigureS3a.pdf \
../figures/FigureS3a.png
```


```{r manhattanGenotype_png, fig.cap = "*Figure 3a: Manhattan plot showing results for all SNPs on chromosomes 1:21 for analysis by genotype. The horizontal line indicates the cutoff for an FDR of 5%, with SNPs with a Bonferroni-adjusted p-value < 0,05 additionally shown in green*", out.width='80%'}
knitr::include_graphics(file.path("..", "figures", "FigureS3a.png"))
```

```{r plotSigGenotypes, fig.cap="*Variants detected as significant to an FDR of 5%. Those denoted with an asterisk received a Bonferroni-adjusted p-value < 0.05 and these typically involved reduction of the minor allele frequency and a shift towards homozygous reference. The remaining sites showed a combination of the same and increasing minor allele abundance, with some variants becoming exclusively heterozygous in the 2012 population.*"}
genotypeResults %>% 
  filter(FDR < alpha) %>% 
  left_join(
    filter(
      genotypes, 
      snpID %in% .$snpID, 
      Population %in% c(1996, 2012)
    )
  ) %>% 
  mutate(
    Genotype = c("AA", "AB", "BB")[Genotype + 1],
    Genotype = as.factor(Genotype),
    Population = as.factor(Population),
    snpID = case_when(
      adjP < alpha ~ paste0(snpID, "*"),
      adjP >= alpha ~ as.character(snpID)
    ),
    snpID = factor(snpID, levels = unique(snpID))
  ) %>%
  ggplot(aes(Genotype, fill = Population)) + 
  geom_bar(position = position_dodge2(preserve = "single")) + 
  facet_wrap(~snpID, drop = FALSE, ncol = 7) + 
  scale_fill_manual(values = c("black", "blue")) +
  labs(y = "Count") + 
  theme(
    legend.direction = "horizontal", 
    legend.position = c(0.75, 0.05)
  )
```

```{r exportGenotypeRes}
write_tsv(genotypeResults, "genotypeResults.tsv")
```



### Allele Frequency Model

This model tests:  
H<sub>0</sub>: No association between allele frequencies and populations  
H<sub>A</sub>: An association exists between allele frequencies  and populations

```{r alleleResults}
alleleResults <- genotypes %>%
  filter(
    Population != 2010,
    !variant.id %in% regionSNPs$variant.id
  ) %>%
  group_by(snpID, Population) %>%
  summarise(
    P = sum(2 - Genotype),
    Q = sum(Genotype)
  ) %>%
  ungroup() %>%
  split(f = .$snpID) %>%
  mclapply(function(x){
    m <- as.matrix(x[c("P", "Q")])
    ft <- list(p.value = NA) 
    if (length(m) == 4) {
      ft <- fisher.test(m)
    }
    x %>%
      mutate(MAF = Q / (P + Q)) %>%
      dplyr::select(snpID, Population, MAF) %>%
      spread(Population, MAF) %>%
      mutate(p = ft$p.value)
  },mc.cores = mc) %>%
  bind_rows() %>%
  filter(!is.na(p)) %>%
  dplyr::rename(
    MAF_1996 = `1996`,
    MAF_2012 = `2012`
  ) %>%
  mutate(
    FDR = p.adjust(p, "fdr"),
    adjP = p.adjust(p, "bonferroni"),
    # Make the MAF is relative to the 1996 population not the reference
    MAF_2012 = case_when(
      MAF_1996 > 0.5 ~ 1 - MAF_2012,
      MAF_1996 <= 0.5 ~ MAF_2012
    ),
    MAF_1996 = case_when(
      MAF_1996 > 0.5 ~ 1 - MAF_1996,
      MAF_1996 <= 0.5 ~ MAF_1996
    )
  ) %>%
  arrange(p) %>%
  left_join(
    genotypes %>% distinct(snpID, chromosome, position)
  ) %>%
  dplyr::select(snpID, chromosome, position, everything())
```


```{r}
alleleSnps <- alleleResults %>%
  dplyr::filter(FDR < alpha) %>%
  .[["snpID"]]
```


```{r}
snpsGR %>% 
  subset(snpID %in% alleleSnps) %>% 
  join_nearest(ensGenes) %>%
  as.data.frame() %>%
  left_join(
    ensGenes %>% as.data.frame(),
    by = c("Name", "gene_id", "seqnames", "description"),
    suffix = c("_snp", "_gene")
  ) %>%
  as_tibble() %>%
  dplyr::select(seqnames, start_snp, variant.id, snpID, gene_id, Name, start_gene, end_gene) %>%
  mutate(dist2Gene = case_when(
    start_snp >= start_gene & start_snp <= end_gene ~ 0L,
    start_snp < start_gene ~ start_gene - start_snp,
    start_snp > end_gene ~ start_snp - end_gene
  ),
  Name = case_when(
    dist2Gene > maxKb*1000 ~ "",
    dist2Gene < maxKb*1000 ~ Name
  ),
    gene_id = case_when(
    dist2Gene > maxKb*1000 ~ "",
    dist2Gene < maxKb*1000 ~ gene_id
  ),
  inExon = snpID %in% ensExons$snpID,
  dist2Gene = case_when(
    Name == "" ~ "",
    inExon ~ "CDS",
    dist2Gene == 0 & !inExon ~ "Intronic",
    dist2Gene > 0 ~ comma(dist2Gene)
  ),
  seqnames = as.character(seqnames)
  ) %>%
  dplyr::select(
    snpID,
    chromosome = seqnames,
    position = start_snp,
    gene_id, Name, dist2Gene
  ) %>%
  left_join(alleleResults) %>%
  dplyr::select(
    snpID, chromosome, position, starts_with("MAF"),
    gene_id, Name, dist2Gene, p, FDR, adjP
  ) %>%
  arrange(p) %>%
  mutate(
    position = comma(position),
    MAF_1996 = sprintf("%.3f", MAF_1996),
    MAF_2012 = sprintf("%.3f", MAF_2012),
    p = tidyP(p),
    FDR = tidyP(FDR),
    adjP = tidyP(adjP)
  ) %>%
  pander(
    justify = "llrrrlllrrr",
    style = "rmarkdown",
     big.mark = ",",
    split.tables = Inf,
      caption = paste(
        "*SNPs considered as significant when analysing by genotype using an FDR cutoff of", 
        percent_format(1)(alpha),
        "Minor allele frequencies are reported based on frequencies in the 1996 population.",
        "Any genes within", paste0(maxKb, "kb"), "are also shown.",
        "Bonferroni-adjusted p-values are given in the final column.*"
      )
  )
```


Under this model:

- `r nrow(filter(alleleResults, adjP < alpha))` SNP alleles were detected as being significantly associated with the two populations when controlling the FWER at &#945; = `r alpha`.
- `r nrow(filter(alleleResults, FDR < alpha))` SNP alleles were detected as being significantly associated with the two populations when controlling the FDR at &#945; = `r alpha`


```{r manhattanAllele, fig.height=5, fig.width=8, include = FALSE}
pdf(file.path("..", "figures", "FigureS3b.pdf"), height = 0.6*h, width = w)
hLine <- max(filter(alleleResults, FDR < alpha)$p)
hLight <- as.character(filter(alleleResults, adjP < 0.05, chromosome %in% 1:21)$snpID)
alleleResults %>%
  mutate(Chr = as.numeric(chromosome)) %>%
  filter(!is.na(Chr)) %>%
  dplyr::select(SNP = snpID, CHR = Chr, BP = position, P = p) %>%
  manhattan(
    suggestiveline = FALSE, 
    genomewideline = -log10(hLine),
    highlight = hLight
  )
dev.off()
```

```{bash}
convert \
-density 600 \
-size 1920 \
../figures/FigureS3b.pdf \
../figures/FigureS3b.png
```


```{r manhattanAllele_png, fig.cap = "Manhattan plot showing results for all SNPs on chromosomes 1:21 when analysing by allele frequencies. The horizontal line indicates the cutoff for an FDR of 5%, with SNPs considered significant under the Bonferroni adjustment shown in green.", out.width='80%'}
knitr::include_graphics(file.path("..", "figures", "FigureS3b.png"))
```


```{r plotMAF, fig.cap = "*Comparison of minor allele frequencies in both populations. SNPs with an FDR-adjusted p-value < 0.05 are coloured red, whilst those with a Bonferroni-adjusted p-value are additionally labelled.*"}
alleleResults %>%
    mutate(Sig = FDR < alpha) %>%
    ggplot(aes(MAF_1996, MAF_2012, colour = Sig, label = snpID)) +
    geom_point() +
    geom_text_repel(data = . %>% filter(adjP < alpha)) +
    # geom_text_repel(data = . %>%filter(FDR < alpha & MAF_2012 > MAF_1996)) +
    scale_colour_manual(values = c(rgb(0.5, 0.5, 0.5, 0.6), "red")) +
    labs(x = "MAF (1996)", y = "MAF (2012)") +
    theme(legend.position = "none")
```

```{r plotSigAlleles, fig.cap="*Variants detected as significant to an FDR of 5% under the Allele Frequency model. Those denoted with an asterisk received a Bonferroni-adjusted p-value < 0.05 and these typically involved reduction of the minor allele frequency and a shift towards homozygous reference. The remaining sites showed a combination of the same and increasing minor allele abundance.*"}
alleleResults %>% 
  filter(FDR < alpha) %>% 
  gather(Population, B, contains("MAF")) %>%
  mutate(
    Population = str_remove(Population, "MAF_"),
    A = 1 - B,
    Population = as.factor(Population),
    snpID = case_when(
      adjP < alpha ~ paste0(snpID, "*"),
      adjP >= alpha ~ as.character(snpID)
    ),
    snpID = factor(snpID, levels = unique(snpID))
  ) %>% 
  gather(Allele, Freq, A, B) %>%
  ggplot(aes(Allele, y = Freq, fill = Population)) + 
  geom_bar(
    position = position_dodge2(preserve = "single"),
    stat = "identity"
  ) + 
  facet_wrap(~snpID, drop = FALSE, ncol = 7) + 
  scale_fill_manual(values = c("black", "blue")) +
  labs(y = "Count") + 
  theme(
    legend.direction = "horizontal", 
    legend.position = c(0.75, 0.05)
  )
```


```{r exportAlleleResults}
write_tsv(alleleResults, "alleleResults.tsv")
```


### Analysis Using the FLK model

This was performed separately and no unique SNPs were detected in addition to any detected above. 

### Export of Data for Bayescan

A VCF was required with only the 1996 and 2012 populations, and restricted to the candidate SNPs after pruning for linkage disequilibrium and detection of the allele in the 1996 population.

```{r var2VCF}
var2VCF <- genotypes %>% 
  distinct(variant.id, snpID) %>%
  filter(snpID %in% filter(regionResults, p > 0.05)$snpID) %>% 
  .[["variant.id"]]
```

```{r, eval=FALSE}
gdsFile <- seqOpen(gdsPath, readonly = TRUE)
seqSetFilter(
  gdsFile, 
  variant.id = var2VCF,
  sample.id = sampleID %>% 
    filter(Population %in% c(1996, 2012)) %>% 
    .[["Sample"]]
)
seqGDS2VCF(gdsFile, "../5_stacks/vcf/filtered.vcf.gz")
seqResetFilter(gdsFile)
seqClose(gdsFile)
```

### Inclusion of Bayescan Results

After running Bayescan, results were imported.
Unfortunately, these results had IDs which were taken from the line in the above VCF and need to be converted back to IDs compatible with previous analysis.


```{r bayRes}
bayRes <- file.path("..", "external", "BayescanOut.txt.gz") %>%
  gzfile() %>%
  read_delim(delim = " ", col_names = FALSE, skip = 1, col_types = "innnnn-") %>%
  set_colnames(c("ID", "prob","log10.PO.","qval","alpha","fst")) %>%
  mutate(variant.id = var2VCF[ID])
```

```{r}
bayRes %>%
  dplyr::select(variant.id, qval, fst) %>%
  dplyr::filter(qval < alpha) %>%
  left_join(
    genotypes %>% 
      distinct(snpID, .keep_all = TRUE) %>% 
      dplyr::select(variant.id, chromosome, position, snpID)
  ) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE,
    ignore.strand = TRUE,
    seqinfo = seqinfo(snp2Gene), 
    seqnames.field = "chromosome",
    start.field = "position",
    end.field = "position"
  ) %>%
  join_nearest(ensGenes) %>%
  as.data.frame() %>%
  left_join(
    ensGenes %>% as.data.frame(),
    by = c("Name", "gene_id", "seqnames", "description"),
    suffix = c("_snp", "_gene")
  ) %>%
  as_tibble() %>%
  dplyr::select(seqnames, start_snp, variant.id, snpID, gene_id, Name, start_gene, end_gene, qval, fst) %>%
  mutate(dist2Gene = case_when(
    start_snp >= start_gene & start_snp <= end_gene ~ 0L,
    start_snp < start_gene ~ start_gene - start_snp,
    start_snp > end_gene ~ start_snp - end_gene
  ),
  Name = case_when(
    dist2Gene > maxKb*1000 ~ "",
    dist2Gene < maxKb*1000 ~ Name
  ),
    gene_id = case_when(
    dist2Gene > maxKb*1000 ~ "",
    dist2Gene < maxKb*1000 ~ gene_id
  ),
  inExon = snpID %in% ensExons$snpID,
  dist2Gene = case_when(
    Name == "" ~ "",
    inExon ~ "CDS",
    dist2Gene == 0 & !inExon ~ "Intronic",
    dist2Gene > 0 ~ comma(dist2Gene)
  ),
  seqnames = as.character(seqnames)
  ) %>%
  dplyr::select(
    snpID, 
    chromosome = seqnames, position = start_snp, 
    gene_id, Name, dist2Gene, fst, qval) %>%
  # Replace the one dropped by `join_nearest()`
  bind_rows(
    bayRes %>%
      dplyr::select(variant.id, qval, fst) %>%
      dplyr::filter(qval < alpha) %>%
      left_join(
        genotypes %>% 
          distinct(snpID, .keep_all = TRUE) %>% 
          dplyr::select(variant.id, chromosome, position, snpID)
      ) %>%
      mutate(
        gene_id = "",
        Name = "",
        dist2Gene = ""
      ) %>%
      dplyr::select(-variant.id)
  ) %>%
  distinct(snpID, .keep_all = TRUE) %>%
  mutate(position = comma(position)) %>%
  arrange(qval) %>%
  mutate(
    fst = tidyP(fst),
    qval = tidyP(qval)
  ) %>%
  pander(
    justify = "llrllrrr",
    style = "rmarkdown",
     big.mark = ",",
    split.tables = Inf,
      caption = paste(
        "*Results from Bayescan with a q-value <", paste0(alpha, "."),
        "The nearest gene within", paste0(maxKb, "kb"), "is also given.",
        "One SNP detected by Bayescan was located on a scaffold with no genes (GL019119).*"
      )
  )
```


```{r plotVenn, fig.cap = "*Summary of results from all three methods, using an FDR of 0.05 for each approach. All SNPs detected under Bayescan were identified using Fisher's Exact Test for both allele and genotype analysis*"}
grid.newpage()
list(
  Genotype = dplyr::filter(genotypeResults, FDR < alpha)$variant.id,
  Allele = alleleResults %>%
    left_join(snps) %>%
    dplyr::filter(FDR < alpha) %>%
    .[["variant.id"]],
  Bayescan = dplyr::filter(bayRes, qval < 0.05)$variant.id
) %>% 
 VennDiagram::venn.diagram(filename = NULL) %>%
  grid.draw()
```

```{r sigSnps}
sigSnps <- c(
  filter(alleleResults, FDR < alpha)$snpID,
  filter(genotypeResults, FDR < alpha)$snpID
) %>%
  unique()
```


```{r}
suppressWarnings(
  snpsGR %>%
    subset(snpID %in% sigSnps) %>%
    resize(width = 1000*maxKb*2, fix = "center") %>%
    trim() %>%
    find_overlaps(ensGenes) %>%
    as.data.frame() %>%
    as_tibble() %>%
    left_join(snps, by = c("variant.id", "snpID", "Col")) %>%
    dplyr::select(snpID, gene_id, Name, chromosome, position) %>%
    left_join(
      as.data.frame(ensGenes), by = c("gene_id", "Name")
    ) %>%
    mutate(
      distance = case_when(
        position < start ~ comma(start - position, accuracy = 1),
        position > end ~ comma(position - end, accuracy = 1),
        snpID %in% ensExons$snpID ~ "Exonic",
        position >= start & position <= end ~ "Intronic"
      ),
      model = ifelse(
        snpID %in% filter(alleleResults, FDR < alpha)$snpID, "A", ""
        ),
      model = ifelse(
         snpID %in% dplyr::filter(snps, variant.id %in% dplyr::filter(bayRes, qval < 0.05)$variant.id)$snpID,
         paste0(model, "B"), model
      ),
      model = ifelse(
        snpID %in% filter(genotypeResults, FDR < alpha)$snpID,
        paste0(model, "G"), model
      ),
      width = comma(width, accuracy = 1)
    ) %>%
    unite(location, chromosome, position, sep = ":") %>%
    dplyr::select(
      `SNP Location` = location, 
      `SNP ID` = snpID, 
      `Gene ID` = gene_id, 
      `Gene Name` = Name, 
      `Gene Width` = width, 
      `Distance` = distance, 
      Models = model
    )
) %>% 
  pander(
    justify = "llllrll",
    style = "rmarkdown",
    split.tables = Inf,
    # big.mark = ",",
    caption = paste0(
      "*All putatively significant SNPs with genes located within ", maxKb, "kb. ",
      "The distance from SNP to gene is indicated, unless the SNP is intronic/exonic. ",
      "The model(s) which identified the SNP are indicated in the final column ",
      "as (A) the Allele Model; (B) Bayescan and (G) the Genotype Model.*"
    )
  )
```



## Enrichment testing

```{r fullPaths}
fullPaths <- list(
  toTable(GOBPANCESTOR) %>% set_names(c("go_id", "ancestor")),
  toTable(GOCCANCESTOR) %>% set_names(c("go_id", "ancestor")),
  toTable(GOMFANCESTOR) %>% set_names(c("go_id", "ancestor"))
) %>% 
  bind_rows() %>%
  dplyr::filter(!ancestor %in% c("all", "GO:0008150", "GO:0003674", "GO:0005575")) %>%
  as_tibble() %>%
  bind_rows(
    tibble(
      go_id = unique(.$go_id),
      ancestor = unique(.$go_id)
    )
  ) %>%
  arrange(go_id, ancestor)
```

```{r goSummaries}
goSummaries <- url("https://uofabioinformaticshub.github.io/summaries2GO/data/goSummaries.RDS") %>%
  readRDS() %>%
  dplyr::rename(go_id = id)
```

### GO Analysis By Gene

For the initial GO analysis, any gene within `r maxKb`kb of a SNP considered as significant were tested for any enrichment of biological characteristics.

```{r allSnpsGR}
allSnpsGR <- snps %>%
  dplyr::select(chromosome, position, snpID) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE,
    seqnames.field = "chromosome", 
    start.field = "position", 
    end.field = "position", 
    seqinfo = seqinfo(ensGenes)) %>%
  join_overlap_inner(
    ensGenes,
    maxgap = maxKb*1000
  )
```

```{r sigSnpGR}
sigSnpGR <- allSnpsGR %>%
  subset(snpID %in% sigSnps)
```


```{r nGenes}
nGenes <- allSnpsGR$gene_id %>%
  unique() %>%
  length()
nSigGenes <- sigSnpGR$gene_id %>%
  unique() %>%
  length()
```

`r nGenes` genes were identified as being within `r maxKb`kb of a SNP.
For the `r length(unique(sigSnpGR$snpID))` candidate SNPs within this distance of a gene, `r nSigGenes` genes were identified, and these were analysed for enrichment of any GO terms.

Mappings from genes to GO terms were downloaded manually from the biomart server at www.ensembl.org.

```{r bm}
bm <- file.path("..", "external", "gene2GO_biomart.tsv.gz") %>%
  gzfile() %>%
  read_tsv() %>%
  dplyr::filter(!is.na(`GO domain`)) %>%
  dplyr::rename(gene_id = `Gene stable ID`,
         ontology = `GO domain`,
         go_id = `GO term accession`,
         go_term = `GO term name`) %>%
  dplyr::filter(gene_id %in% allSnpsGR$gene_id) %>%
  left_join(fullPaths) %>%
  dplyr::select(gene_id, go_id = ancestor) %>%
  distinct(gene_id, go_id) %>%
  dplyr::filter(!is.na(go_id))
```

Complete paths of GO terms back to ontology roots were obtained for each gene.
The minimum number of steps back to the ontology roots were also obtained for each GO ID.


```{r goRes}
goRes <- bm %>%
  mutate(inSig = gene_id %in% sigSnpGR$gene_id) %>%
  group_by(go_id) %>%
  summarise(N = dplyr::n(), 
            nSig = sum(inSig)) %>%
  left_join(goSummaries) %>%
  filter(nSig > 0, shortest_path > 2) %>%
  distinct(go_id, .keep_all = TRUE) %>%
  split(f = .$go_id) %>%
  mclapply(function(x){
    ft <- matrix(
      c(nSigGenes - x$nSig, x$nSig, nGenes - nSigGenes - x$N, x$N),
      nrow = 2, byrow = TRUE) %>%
      fisher.test()
    mutate(x, p = ft$p.value)
  }, mc.cores = mc) %>%
  bind_rows() %>%
  mutate(Expected = nSigGenes * N / nGenes,
         FDR = p.adjust(p, "fdr"),
         adjP = p.adjust(p, "bonferroni"),
         term = Term(go_id),
         ontology = Ontology(go_id)) %>%
  dplyr::select(go_id, term, ontology, N, nSig, Expected, p, adjP, FDR) %>%
  arrange(p)
```


```{r}
goRes %>%
  dplyr::filter(FDR == min(FDR), nSig > 1) %>%
  mutate(
    Term = Term(go_id),
    Genes  = vapply(go_id, function(x){
      subset(sigSnpGR, gene_id %in% I(
        dplyr::filter(bm, go_id == x) %>%
          .[["gene_id"]]))$Name %>%
        sort() %>%
        unique() %>%
        paste(collapse = ", ")
    }, character(1))
  ) %>%
  dplyr::select(ID = go_id, Ontology = ontology, Term, Total = N, Sig = nSig, Expected, p, FDR, Genes) %>%
  mutate(
    ID = gsub(":", "\\\\:", ID),
    Expected = tidyP(Expected),
    p = tidyP(p),
    FDR = round(FDR, 3)
  ) %>%
  distinct(Ontology, Genes, .keep_all = TRUE) %>%
  pander(
    split.tables = Inf, style = "rmarkdown", 
    justify = "lllrrrrrl",
    caption = paste(
      "*Most highly ranked GO terms.", 
      "Where multiple terms resulted from the identical combination of genes, only the most highly ranked terms is shown.",
      "Overall, this corresponds to an FDR of", 
      percent(max(.$FDR)), ".*"
    )
  )
```

## Meta Analysis With Previous Paper

The results in the Turretfield population from the paper [*Resistance to RHD virus in wild Australia rabbits: Comparison of susceptible and resistant individuals using a genomewide approach*](https://doi.org/10.1111/mec.14228) were then compared to these results.
In this previous analysis, results were generated by using Fisher's Exact Test to test genotypes for association with RHDV susceptibility or resistance.

```{r tf, cache=TRUE}
tfGR <- url("https://raw.githubusercontent.com/hdetering/orycun/master/results/filteredSNPScores.csv") %>% 
  read_csv() %>%
  distinct(min_snp, .keep_all = TRUE) %>%
  dplyr::select(chromosome = chr_name, pos, snpID = min_snp, p, FDR) %>%
  mutate(chromosome = gsub("^chr", "", chromosome)) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE,
    ignore.strand = TRUE, 
    seqinfo = seqinfo(ensGenes),
    seqnames.field = "chromosome", 
    start.field = "pos", 
    end.field = "pos") %>%
  sort()
```

Of the `r comma(length(tfGR))` unique SNPs contained in the previous dataset, only `r tfGR %>% find_overlaps(allSnpsGR) %>% reduce() %>% length() %>% comma()` were also identified in the current dataset.
A meta-analysis was performed on these SNPs using the results from both separate analyses which used Fisher's Exact Test on genotypes.
In this approach, Fisher's method was used to combine p-values, and resultant p-values were drawn from a $\chi^2$ distribution with 4 degrees of freedom (i.e. $2k$).

```{r combinedResults}
combinedResults <- tfGR %>%
  find_overlaps(allSnpsGR) %>%
  mcols() %>%
  as.data.frame() %>%
  as_tibble() %>%
  dplyr::select(snpID = snpID.y, tf.p = p) %>%
  distinct(snpID, .keep_all = TRUE) %>%
  left_join(
    genotypeResults %>%
      dplyr::select(snpID, chromosome, position, geno.p = p)) %>% 
  mutate(
    X = -2*(log(tf.p) + log(geno.p)), 
    p = pchisq(X, 4, lower.tail = FALSE), 
    adjP = p.adjust(p, "bonferroni"),
    FDR = p.adjust(p, "fdr")) %>% 
  arrange(p) 
```

Using Bonferroni's adjustment to control the FWER at 0.05 gave `r nrow(filter(combinedResults, adjP < 0.05))` candidate SNPs, whilst using the more relaxed criterion of an FDR < 0.05 gave `r nrow(filter(combinedResults, FDR < 0.05))` candidate SNPs.

```{r}
combinedResults %>%
  dplyr::filter(FDR < 0.05) %>%
  dplyr::select(
    snpID,
    Chromosome = chromosome, 
    Position = position,
    p, adjP, FDR
  ) %>%
  mutate(
    p = tidyP(p),
    adjP = tidyP(adjP),
    FDR = tidyP(FDR)
  ) %>%
  pander(
    justify = "lrrrrr",
    caption = "*Significant SNPs from the meta-analysis using the previous published results from a separate population, using an FDR < 0.05 as the sole criterion for significance.*"
    )
```


```{r}
allSnpsGR %>% 
  subset(snpID %in% dplyr::filter(combinedResults, FDR < 0.05)$snpID) %>%
  subset(!is.na(Name)) %>% 
  as.data.frame() %>%
  dplyr::select(
    snpID, 
    Chromosome = seqnames,
    BP = start, 
    ID = gene_id,
    Name,
    Description = description
  ) %>%
  mutate(Description = str_remove(Description, " \\[Source.+")) %>%
  pander(
    style = "rmarkdown", split.table = Inf, justify = "lrrlll",
    caption = paste0(
      "*Genes within ", maxKb, "kb of SNPs identified in the meta-analysis.*"
    )
  )
```

```{r, fig.cap="*Comparison of p-values for SNPs detected in both analyses, with those considered as significant after meta-analysis shown in red.*"}
combinedResults %>%
  mutate(Sig = FDR < 0.05) %>%
  ggplot(aes(-log10(geno.p), -log10(tf.p), colour = Sig)) +
  geom_point() +
  geom_abline(slope =1, intercept = 0, linetype = 2, colour = "blue") +
  geom_text_repel(aes(label = snpID),
                  data = . %>% filter(Sig)) +
  scale_color_manual(values = c("black", "red")) +
  labs(x = expression(paste(-log[10], "p (Genotype Analysis)")),
       y = expression(paste(-log[10], "p (Previous Analysis)"))) +
  theme(legend.position = "none")
```


```{r, fig.cap = "*Allele distributions in the 1996 and 2012 populations for SNPs identified under the meta-analysis.*"}
combinedResults %>%
  dplyr::filter(FDR < 0.05) %>%
  dplyr::select(snpID) %>%
  left_join(genotypes) %>%
  filter(Population != 2010,
         !variant.id %in% regionSNPs$variant.id) %>%
  mutate(Genotype = as.factor(Genotype),
         Population = as.factor(Population)) %>%
    ggplot(aes(Genotype, fill = Population)) + 
    geom_bar(position = position_dodge2(preserve = "single")) + 
    facet_wrap(~snpID, drop = FALSE) + 
    scale_fill_manual(values = c("black", "blue")) +
    labs(y = "Count") + 
    theme(
        legend.direction = "horizontal", 
        legend.position = c(0.75, 0.05)
    )
```


### GO Enrichment

This list was also used to test for GO enrichment.

```{r setupMetaGO}
sigSnpGR <- allSnpsGR %>%
  subset(snpID %in% filter(combinedResults, FDR < 0.05)$snpID)
nGenes <- allSnpsGR %>%
  subset(snpID %in% combinedResults$snpID) %>%
  unique() %>%
  length()
nSigGenes <- sigSnpGR$gene_id %>%
  unique() %>%
  length()
```

```{r combinedGoRes}
combinedGoRes <- bm %>%
  dplyr::filter(
    gene_id %in% subset(allSnpsGR, snpID %in% combinedResults$snpID)$gene_id) %>%
  mutate(inSig = gene_id %in% sigSnpGR$gene_id) %>%
  group_by(go_id) %>%
  summarise(N = dplyr::n(), 
            nSig = sum(inSig)) %>%
  left_join(goSummaries) %>%
  filter(nSig > 0, shortest_path > 2) %>%
  distinct(go_id, .keep_all = TRUE) %>%
  split(f = .$go_id) %>%
  mclapply(function(x){
    ft <- matrix(
      c(nSigGenes - x$nSig, x$nSig, nGenes - nSigGenes - x$N, x$N),
      nrow = 2, byrow = TRUE) %>%
      fisher.test()
    mutate(x, p = ft$p.value)
  }, mc.cores = mc) %>%
  bind_rows() %>%
  mutate(Expected = nSigGenes * N / nGenes,
         FDR = p.adjust(p, "fdr"),
         adjP = p.adjust(p, "bonferroni"),
         term = Term(go_id),
         ontology = Ontology(go_id)) %>%
  dplyr::select(go_id, term, ontology, shortest_path, N, nSig, Expected, p, adjP, FDR) %>%
  arrange(p)
```

Inspection of the results revealed an enrichment for terms related to haemoglobin, however, as these genes are co-located this was revealed to be due to a single SNP with multiple haemoglobin genes within 40kb and can be disregarded. 

```{r}
combinedGoRes %>%
  dplyr::filter(
    nSig > Expected,
    FDR < 0.05
  ) %>%
  mutate(Term = Term(go_id),
         Genes  = vapply(go_id, function(x){
           subset(sigSnpGR, gene_id %in% I(
             dplyr::filter(bm, go_id == x) %>%
               .[["gene_id"]]))$Name %>%
             sort() %>%
             unique() %>%
             paste(collapse = ", ")
         }, character(1)),
         nSnps = vapply(go_id, function(x){
           subset(sigSnpGR, gene_id %in% I(
             dplyr::filter(bm, go_id == x) %>%
               .[["gene_id"]]))$snpID %>%
             unique() %>%
             length()
           }, integer(1))) %>%
  dplyr::select(ID = go_id, Ontology = ontology, Term, Total = N, Sig = nSig, nSnps, Expected, p, FDR, Genes) %>%
  mutate(ID = gsub(":", "\\\\:", ID),
         Expected = round(Expected, 3),
         p = round(p, 4),
         FDR = round(FDR, 3)) %>%
  distinct(Ontology, Genes, .keep_all = TRUE) %>%
  pander(
    split.tables = Inf, style = "rmarkdown", 
    justify = "lllrrrrrrl",
    caption = paste(
      "Most highly ranked GO terms for combined results.", 
      "Where multiple terms resulted from the identical combination of genes, only the most highly ranked terms is shown.",
      "Overall, this corresponds to an FDR of", 
      percent(max(.$FDR)), ".*"
    )
  )
```


## Session Information

```{r sessionInfo, echo=FALSE, results='asis'}
pander(sessionInfo()) 
```

```{r deleteVennLogs, echo=FALSE, results='hide'}
list.files(pattern = "log$") %>% file.remove()
```

