---
title: "Linkage Analysis"
author: "Steve Pederson"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document: 
    fig_caption: yes
    fig_height: 8
    fig_width: 10
    toc: true
    toc_float: true
    code_folding: hide 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE, warning = FALSE,
  fig.align = "center"
)
```


```{r loadPackages}
library(SeqArray)
library(SNPRelate)
library(pander)
library(scales)
library(magrittr)
library(tidyverse)
library(parallel)
library(plyranges)
library(rtracklayer)
```

```{r setOptions}
if (interactive()) here::here("R") %>% setwd()
theme_set(theme_bw())
panderOptions('big.mark', ",")
panderOptions('table.style', "rmarkdown")
panderOptions('table.split.table', Inf)
panderOptions("missing", "")
mc <- min(12, detectCores() - 1)
```

## Linkage Analysis

```{r loadGDS}
gdsPath <- file.path("..", "5_stacks", "gds", "populations.snps.gds")
gdsFile <- seqOpen(gdsPath, readonly = TRUE)
```

```{r sampleID}
sampleID <- tibble(
  Sample = seqGetData(gdsFile, "sample.annotation/Sample"),
  Population = seqGetData(gdsFile, "sample.annotation/Population"),
  Location = seqGetData(gdsFile, "sample.annotation/Location")
) %>%
  filter(Population != 2010)
```

Autosomal SNP IDs for the 1996 and 2012 samples were loaded.

```{r snps, include=FALSE}
seqSetFilter(gdsFile, sample.id = sampleID$Sample)
snps <- tibble(
  variant.id = seqGetData(gdsFile, "variant.id") ,
  chromosome = seqGetData(gdsFile, "chromosome"),
  position = seqGetData(gdsFile, "position"),
  snpID = seqGetData(gdsFile, "annotation/id"),
  N = seqGetData(gdsFile, "genotype") %>% 
    apply(3, colSums) %>% 
    apply(2, function(x){sum(!is.na(x))}) 
) %>%
  mutate(snpID = str_remove(snpID, ":.$"),
         snpID = str_replace(snpID, ":", "_")) %>%
  separate(snpID, into = c("Locus ID", "Col"), remove = FALSE) %>%
  mutate_at(vars(`Locus ID`, Col), as.integer) %>%
  mutate(Col = Col - 1,
         snpID = paste(`Locus ID`, Col, sep = "_")) 
seqResetFilter(gdsFile)
```

To define the SNPs used for estimation of Linkage Decay, select the 10,000 autosomal SBPs identified in the highest number of idividuals.
Here this was set as SNPs identified in >95% of individuals (i.e. > 104).
For simpler plotting and estimation of decay, only SNPs < 1Mb apart were retained.


```{r allLD, results='hide', include=FALSE, cache=TRUE}
seqSetFilter(gdsFile, sample.id = sampleID$Sample)
allLD <- snps %>% 
  filter(
    N >= 105, # Only the SNPs identified in > 95% of samples
    chromosome %in% 1:22
  ) %>%
  arrange(chromosome, position) %>%
  split(f = .$chromosome) %>%
  lapply(function(x){
    chr <- unique(x$chromosome)
    # Obtain the LD matrix
    ld <- snpgdsLDMat(
      gdsobj = gdsFile, 
      snp.id = x$variant.id,
      slide = 0,
      num.thread = mc,
      with.id = TRUE)
    ld$LD %>%
      .[lower.tri(.)] %>%
      enframe(value = "LD") %>%
      dplyr::select(LD) %>%
      cbind(
        x$variant.id %>%
          combn(2) %>%
          t() %>%
          set_colnames(c("ID1", "ID2"))
      ) %>%
      as_tibble() %>%
      left_join(x %>% dplyr::select(ID1 = variant.id, Pos1 = position)) %>%
      left_join(x %>% dplyr::select(ID2 = variant.id, Pos2 = position)) %>%
      mutate(bp = Pos2 - Pos1,
             chromosome = chr) %>%
      dplyr::filter(!is.na(LD),bp < 1e6)
  }) %>%
  bind_rows()
seqResetFilter(gdsFile)
n2plot <- 1e5
```

A function was then defined to fit Hill & Weir's linkage decay equation using a non-linear fit.

```{r hwDecay}
hwDecay <- function(d, r2, n = nrow(sampleID), k = 1e3){
  st <- c(C = 0.1)
  fit <- nls(
    r2 ~ ((10 + C*d) / ((2 + C*d)*(11 + C*d))) *
      (1 +((3 + C*d)*(12 + 12*C*d + (C*d)^2)) / (n*(2 + C*d)*(11 + C*d))),
    start = st,
    control = nls.control(maxiter = 100)
  )
  rho <- summary(fit)$parameters[1]
  x <- seq(0, max(d), length.out = k)
  y <- ((10 + rho*x)/((2 + rho*x)*(11 + rho*x))) *
    (1 + ((3 + rho*x)*(12 + 12*rho*x + (rho*x)^2)) / (n*(2 + rho*x)*(11 + rho*x)))
  list(line = tibble(x = x, y = y),
       fit = fit, 
       rho = rho)
}
```

```{r fitDecay}
fitDecay <- allLD %>%
  mutate(r2 = LD^2) %>%
  with(hwDecay(bp, r2))
```


```{r plotDecay, cache=TRUE, dependson='allLD', fig.cap = paste("*Linkage decay for distances up to 1Mb. The Hill-Weir decay line is shown in green as is the distance 40kb. The red dashed line indicates the distances where linkage falls to half of the maximal linkage.", comma(n2plot), "randomly chosen pairs from the complete set of", comma(nrow(allLD)), "pairwise comparisons are shown.*")}
set.seed(1764)
allLD %>%
  sample_n(n2plot) %>%
  ggplot(aes(bp / 1e3, LD^2)) +
  geom_point(alpha = 0.1) +
  geom_line(
    aes(x / 1e3, y), 
    data = fitDecay$line,
    colour = "green"
  ) +
  geom_vline(xintercept = 40, colour = "green", linetype = 2) +
  geom_vline(
    xintercept = dplyr::filter(fitDecay$line, y < 0.5*max(y))$x[1]/1e3,
    colour = "red", 
    linetype = 2
  ) + 
  geom_label(
    x = dplyr::filter(fitDecay$line, y < 0.5*max(y))$x[1]/1e3,
    y = 0.95,
    label = paste(
      round(dplyr::filter(fitDecay$line, y < 0.5*max(y))$x[1]/1e3, 0),
      "Kb"
    ),
    colour = "red"
  ) +
  labs(x = "Distance (Kb)", y = expression(r^2)) +
  scale_x_continuous(
    expand = expand_scale(0.02), 
    breaks = c(40, seq(0, 1e3, by = 200)), 
    minor_breaks = seq(0, 1e3, by = 100)
  ) +
  theme(
    axis.title.y = element_text(angle = 0, hjust = 0.5, vjust = 0.5),
    text = element_text(size = 14)
  )
```

## Analysis of Significant SNPs

```{r d}
d <- 4e4
```

As the set of SNPs under analysis had been pruned in order to only include one representative SNP in the case of linkage, any SNPs in linkage to the detected SNPs were explored, within `r d/1000`kb.
As phasing was not possible outside of a given Stacks Locus, correlations between genotypes were used as a proxy for linkage.
Correlations between were calculated for each timepoint separately as they may not be expected to be the same at each timepoint, given the selctive bottleneck applied by RHDV.

```{r linkageSnpSetup}
alleleResults <- read_tsv("alleleResults.tsv")
genotypeResults <- read_tsv("genotypeResults.tsv")
```

```{r ensGenes, cache=TRUE}
ensGenes <- file.path(
  "..", "external", "Oryctolagus_cuniculus.OryCun2.0.96.gff3.gz"
) %>%
  import.gff3(feature.type = "gene", sequenceRegionsAsSeqinfo = TRUE) %>%
  .[,c("gene_id", "Name", "description")]
```


```{r snpsGR}
snpsGR <- makeGRangesFromDataFrame(
  df = snps, 
  keep.extra.columns = TRUE, 
  ignore.strand = TRUE, 
  seqinfo = seqinfo(ensGenes), 
  seqnames.field = "chromosome", 
  start.field = "position", 
  end.field = "position"
)
```

```{r ensExons}
ensExons <- file.path("..", "external", "Oryctolagus_cuniculus.OryCun2.0.96.gff3.gz") %>%
    import.gff3(feature.type = "exon", sequenceRegionsAsSeqinfo = TRUE) %>%
  join_overlap_inner_within(
    ensGenes, maxgap = 0, minoverlap = 0, suffix = c(".exon", "")
  ) %>%
  join_overlap_inner(snpsGR)
ensExons <- ensExons[,setdiff(colnames(mcols(ensExons)), "Name.exon")]
```


```{r sigRanges}
sigSNPs <- snpsGR %>%
  subset(
    snpID %in% (
      list(
        allele = alleleResults,
        genotype = genotypeResults
      ) %>%
        lapply(dplyr::filter, FDR < 0.05) %>%
        lapply(extract2, "snpID") %>%
        unlist() %>%
        unique()
    )
  ) 
sigRanges <- sigSNPs %>%
  resize(width = 2*d + 1, fix = "center") %>%
  trim()
```

```{r snpsForCor}
snpsForCor <- snpsGR %>%
  filter_by_overlaps(sigRanges) %>%
  join_nearest(sigSNPs) %>%
  split(f = .$snpID.y) %>%
  .[sigSNPs$snpID]
```

```{r snpN, results='hide'}
countN <- function(x){sum(!is.na(x))}
samples <- list(
  ora = dplyr::filter(sampleID, Population == 2012)$Sample,
  gc = dplyr::filter(sampleID, Population == 1996)$Sample
)
snpN <- snpsForCor %>%
  lapply(function(x){
    seqSetFilter(
      gdsFile, 
      sample.id = unlist(samples), 
      variant.id = mcols(x)$variant.id.x
      )
    gt <- seqGetData(gdsFile, "genotype")
    seqResetFilter(gdsFile)
    apply(gt, MARGIN = 3, colSums) %>%
      set_colnames(mcols(x)$snpID.x) %>%
      set_rownames(unlist(samples)) %>%
      as.data.frame() %>%
      rownames_to_column("Sample") %>%
      left_join(sampleID) %>%
      as_tibble() %>%
      group_by(Population) %>%
      summarise_at(vars(contains("_")), countN) %>%
      gather("snpID", "N", -Population) %>%
      spread("Population", "N") %>%
      dplyr::rename(N_1996 = `1996`, N_2012 = `2012`)
  }) %>%
  bind_rows()
```


```{r cors, results='hide'}
cors <- snpsForCor %>%
  lapply(function(x){
    df1 <- snpgdsLDMat(
      gdsFile,
      sample.id = samples$gc,
      snp.id = x$variant.id.x,
      slide = -1,
      method = "corr"
    )$LD %>%
      set_rownames(x$snpID.x) %>%
      set_colnames(x$snpID.x) %>%
      as.data.frame() %>%
      rownames_to_column("snp1") %>%
      gather(key = "snp2", value = "corr1996", -snp1) %>%
      dplyr::filter(
        snp1 %in% sigSNPs$snpID,
        snp1 != snp2
      )
    df2 <- snpgdsLDMat(
      gdsFile,
      sample.id = samples$ora,
      snp.id = x$variant.id.x,
      slide = -1,
      method = "corr"
    )$LD %>%
      set_rownames(x$snpID.x) %>%
      set_colnames(x$snpID.x) %>%
      as.data.frame() %>%
      rownames_to_column("snp1") %>%
      gather(key = "snp2", value = "corr2012", -snp1) %>%
      dplyr::filter(
        snp1 %in% sigSNPs$snpID,
        snp1 != snp2
      )
    left_join(df1, df2)
  }) %>%
  bind_rows() %>%
  as_tibble() %>%
  dplyr::filter(
    !is.nan(corr2012),
    !is.nan(corr1996)
  ) %>%
  mutate(
    LD2_1996 = corr1996^2,
    LD2_2012 = corr2012^2
  ) %>%
  left_join(snps, by = c("snp1" = "snpID")) %>%
  dplyr::select(
    starts_with("snp"), 
    starts_with("corr"),
    starts_with("LD"),
    variant.id, chromosome, 
    pos1 = position, 
    Locus1 = `Locus ID`
  ) %>%
  left_join(
    snps, 
    by = c("snp2" = "snpID", "chromosome" = "chromosome")
  ) %>%
  dplyr::select(
    chromosome,
    starts_with("snp"), 
    starts_with("corr"),
    starts_with("LD"),    
    Locus1, Locus2 = `Locus ID`,
    pos1, pos2 = position
  ) %>%
  dplyr::mutate(
    dist = pos2 - pos1,
    model = case_when(
      snp1 %in% dplyr::filter(alleleResults, FDR < 0.05)$snpID &
        snp1 %in% dplyr::filter(genotypeResults, FDR < 0.05)$snpID ~ "Both",
      snp1 %in% dplyr::filter(alleleResults, FDR < 0.05)$snpID ~ "Allele",
      snp1 %in% dplyr::filter(genotypeResults, FDR < 0.05)$snpID ~ "Genotype" 
    ),
    diff = LD2_2012 - LD2_1996
  )
```

In this analysis, linkage between each significant SNP and those within `r d/1000`kb was investigated.
Under no selective pressure, linkage would be expected to gradually decrease over the passage of time through recombination, as seen below, where correlations (as indicated by the blue regression line) are trending towards zero.
SNPs in strong physical linkage before selective pressure from RHDV tended to remain in strong linkage if within the same locus (~100bp).

```{r plotCors, fig.cap = "*Correlations between pairs of SNPs located within 100kb of the significant SNPs detected in the previous stages. The regression line is shown in blue, with y = x shown in black for reference. SNPs within the same locus are shown in red. Only SNPs identified in >70% of both populations are shown.*"}
labFun <- function(x){1 / x}
cors %>% 
  mutate(`Same Locus` = Locus1 == Locus2) %>%
  dplyr::filter(
    snp1 %in% dplyr::filter(snpN, N_1996 >= 40, N_2012 >= 37)$snpID,
    snp2 %in% dplyr::filter(snpN, N_1996 >= 40, N_2012 >= 37)$snpID    
  ) %>%
  ggplot(aes(corr1996^2, corr2012^2)) +
  geom_point(
    aes(size = 1 / abs(dist), colour = `Same Locus`), 
    alpha = 0.5
  ) +
  geom_abline(slope = 1) + 
  geom_smooth(method = "lm", se = FALSE) +
  scale_size_continuous(
    labels = labFun, 
    breaks = c(1/d, 10^seq(-4, 0)), 
    trans = "sqrt"
  ) + 
  scale_colour_manual(values = c("black", "red")) +
  labs(
    x = expression(paste(r[1996]^2)),
    y = expression(paste(r[2012]^2)),
    size = "Distance (bp)"
  ) 
```


No impact of distance between two SNPs was noted on changes in linkage, however a clear break in linkage was noted for one pair of SNPs

```{r}
cors %>% 
  ggplot(aes(abs(dist)/1000, diff)) +
  geom_point(aes(alpha = LD2_1996)) +
  geom_smooth(method = "lm") +
  labs(
    x = "Distance (kb)",
    y = expression(paste(Delta, r^2))
  ) 
```

One pair of SNPs (2028473_73 & 2028554_103) was noticed as showing a clear break in linkage.
Manual inspection revealed a further SNP (2028643_21; 13:3054236) showing a very similar decrease in linkage (r^2^) from 1 to 0.1361, at a distance of 74.8kb.
Several other unlinked SNPs also appeared in this region, however this observation suggests than an ancestral haplotype from 1996 may have been selected against.
The nearest gene to this region is COP1 (13:3090539-3302219), and the human ortholog (Q8NHY2) is predicted to be located in numerous cellular locations, including the [Golgi Apparatus](https://www.ebi.ac.uk/QuickGO/annotations?geneProductId=Q8NHY2).

```{r}
cors %>% 
  mutate(diff = LD2_2012 - LD2_1996) %>% 
  dplyr::filter(diff == min(diff)) %>%
  dplyr::select(
    Chromosome = chromosome, 
    Position = pos1,
    snp1, snp2, 
    Distance = dist,
    `r~1996~^2^` = LD2_1996, 
    `r~2012~^2^` = LD2_2012
  ) %>%
  pander(
    justify = "lrrrlll",
    caption = "The one SNP pair noted with a clear change in linkage"
  )
```



## Session Info

```{r closeGds, eval=TRUE}
seqClose(gdsFile)
```


```{r sessionInfo, eval=TRUE}
sessionInfo() %>% pander()
```

