---
title: "Simulation of Genetic Drift"
author: "Steve Pederson"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document: 
    fig_caption: yes
    fig_height: 8
    fig_width: 10
    toc: yes
    toc_depth: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE, warning = FALSE)
```


```{r packages}
library(pander)
library(tidyverse)
library(scales)
library(ggrepel)
library(parallel)
if (interactive()) setwd(here::here("R"))
```

```{r setOptions}
theme_set(theme_bw())
panderOptions("table.split.table", Inf)
panderOptions("big.mark", ",")
panderOptions("table.style", "rmarkdown")
```


```{r logit}
logit <- binomial()$linkfun
inv.logit <- binomial()$linkinv
```

```{r pdfParams}
w <- 169/25.4
h <- 169/25.4
```


# Introduction

This analysis takes the filtered SNPs from under analysis and simulates genetic drift under no selective pressure, in order to compare detected changes in allele frequencies to the range of values predicted by the drift model.

## Outline of Drift Simulation

This analysis uses a custom set of scripts collected as an R package and available from https://github.com/steveped/driftSim.
This is able to be installed as below.

```{r install}
remotes::install_github("steveped/driftSim")
library(driftSim)
```

### Initialisation

```{r simParam, cache=TRUE}
nGen <- 16
nSim <- 5000
litterSize <- 20
migRate <- 0.15
alpha <- 0.001
Ne <- c(`1996` = 222, `2012` = 116)
pSurv <- 0.1
f0 <- seq(0.05, 0.5, by = 0.05)
nPops <- c(4, 8)
sigma <- c(0.5, 0.8)
```


As decided the population parameters were defined as:

| Parameter   | Value          | Comment                                                     |
|:----------- |:-------------- |:----------------------------------------------------------- |
| *Ne*~1996~  | `r Ne["1996"]` | Effective Population Size in 1996                           |
| *Ne*~2012~  | `r Ne["2012"]` | Effective Population Size in 2012                           |
| *p*         | `r pSurv`      | Probability of survival for each generation              |
| *g*         | `r nGen`       | The number of generations between 1996 and 2012             |
| *n*         | `r nPops`      | The number of neighbouring populations                      |
| *l*         | `r litterSize` | The annual litter size                                      |
| *r*         | `r migRate`    | The migration rate                                          |
| *f~0~*      | 0.05 to 0.5    | The starting allele frequency in the main population, increased in steps of 0.05  |
| *&sigma;*   | `r sigma`      | The variability around *f~0~* for neighbouring populations |

No selective advantage was specified for any allele.

The simulation parameters were defined as:

| Parameter   | Value     | Comment                                                     |
|:----------- |:--------- |:----------------------------------------------------------- |
| *n*~sim~    | `r nSim`  | The number of simulations at each starting frequency *f~0~* |
| *&alpha;*   | `r alpha` | The significance level for plotting confidence bands        |


To summarise the above:

1. The effective population size was defined as Ne = `r Ne[[1]]` representing the initial population in 1996. 
2. A starting minor allele frequency was selected as one of *f~0~* = `r pander(f0)`. Each rabbit was simulated as heterozygous or homozygous for either the major or minor allele using the initial starting frequency
3. An initial survival rate was defined as p = `r pSurv`, with this functioning as an initial bottleneck, and rabbits were assigned as survivors or fatalities with this probability.
4. This population was considered as the *central population*, and this initialisation process was then repeated **for either `r nPops[1]` or `r nPops[2]` neighbouring populations** of the same size. However, **variability was added to the initial allele frequencies on the logit scale using values of** *&sigma;* = `r pander(sigma)`. The same bottleneck procedure was applied to each of these populations.


# Placing Variability in Context


```{r nSamp}
nSamp <- 1e06
x_0 <- runif(nSamp, 0.5, 0.95)
y_0 <- sigma %>%
  lapply(function(sigma){
    inv.logit(rnorm(nSamp, logit(x_0), sigma))
  })
```


```{r plotAddVars, echo=FALSE, fig.height=5, fig.cap="The effect of adding variability to derive initial frequencies in neighbouring populations. Half of the simulated initial values will be contained by the box for each starting frequency, with the remaining half being outside of these bounds."}
expand.grid(f0 = f0, sigma = sigma) %>%
  apply(
      MARGIN = 1,
      FUN = function(x){
          tibble(
              f0 = x[1], 
              sigma = x[2],
              samples = rnorm(1000, logit(x[1]), sd = x[2])
          ) %>%
              mutate(samples = inv.logit(samples))
      }
  ) %>%
    bind_rows() %>%
    mutate(sigma = paste("sigma ==", sigma)) %>%
    ggplot(aes(x = f0, y = samples)) +
    geom_boxplot(aes(group= as.factor(f0))) +
    facet_wrap(~sigma, scales = "free", labeller = "label_parsed") +
    scale_y_continuous(breaks = seq(0, 1, by = 0.2)) +
    labs(
        x = "Initial Value in Main Population",
        y = "Simulated Initial Values In Other Populations"
    )
```


In order to express these values for *&sigma;* in terms of correlations, `r comma(nSamp)` random values were simulated for a starting frequency (*f~0~*) anywhere between 0.5 and 0.95.
`r comma(nSamp)` _neighbouring population starting values_ were randomly sampled around this using the values of *&sigma;* = `r pander(sigma)`, with the random sampling taking place on the logit scale.

```{r, echo=FALSE, results='asis'}
tibble(
    Sigma = sigma,
    Correlation = vapply(y_0, cor, y = x_0, numeric(1))
) %>%
    pander(
        digits = 2,
        caption = paste("Approximate correlations between starting allele frequencies in two populations, for the chosen values of $\\sigma$")
        )
```


# Run Simulations

```{r allParam}
allParam <- expand.grid(
    f0 = f0,
    mig = migRate,
    n = nPops,
    sd = sigma
) %>%
    mutate(
        N0 = round(Ne["1996"], 0),
        Nt = round(Ne["2012"], 0),
        t = nGen,
        surv = pSurv,
        litter = litterSize
    ) %>%
    split(f = factor(1:nrow(.)))
```

```{r allSim, cache=TRUE, dependson='simParam'}
allSim <- allParam %>%
  mclapply(
      function(x){
          x <- as.list(x)
          replicate(nSim, do.call(simDrift, args = x)$ft)
      },
      mc.cores = 20
  )
```

```{r, echo=FALSE}
write_rds(allSim, "allSim_Oct2019.rds")
```

## Inspection of Simulations

As expected the mean allele frequency after simulation of drift remained approximately equal to the initial starting frequency, however the variability of the final frequency was heavily affected by the number of neighbouring populations and their similarity to the to main population of interest.

```{r simValues, echo=FALSE, fig.cap="Mean and standard devation of allele frequencies ($f_t$) after simulating drift. Polynomial lines of best fit are shown for each value.", fig.height=6}
bind_rows(allParam) %>% 
    mutate(
        n_t = vapply(allSim, mean, numeric(1)),
        sd_t = vapply(allSim, sd, numeric(1)),
        n = as.factor(n),
        sd = as.factor(sd)
    ) %>%
    dplyr::select(f0, nPops = n, sd_0 = sd, n_t, sd_t) %>%
    gather("variable", "value", ends_with("_t")) %>%
    ggplot(aes(f0, value, shape = nPops, colour = sd_0, linetype = nPops)) +
    geom_point( size = 2.5) +
    geom_smooth(method = "lm", formula = y~poly(x,2), se = FALSE, size = 0.5) +
    facet_wrap(
        ~variable, 
        ncol = 2, 
        scales = "free",
        labeller = as_labeller(
            c(n_t = "bar(f[t])", sd_t = "SD(f[t])"), default = label_parsed
        )
    ) +
    labs(
        x = "Initial Value (f0)",
        y = "After Simulations",
        shape = "Number of Neighbouring Populations (n)",
        linetype = "Number of Neighbouring Populations (n)",
        colour = expression(
            paste("Initial Variability in Neighbouring Populations (", sigma, ")")
        )
    ) +
    scale_colour_manual(values = c(rgb(0.1,0.1,0.8), rgb(0.9,0,0))) +
    theme(legend.position = "bottom")
```

- More populations surrounding the main population tended to hold allele frequencies steady over time
- More variability between populations resulted in wider distributions, i.e. more variable allele frequencies, after allowing for drift
- The Gum Creek / Oraparinna population is not an isolated population, but will likely experience migration from a continuous larger population. The variability within this larger population is unknown.
- The most conservative approach to apply in analysis would be to assume a small number of neighbouring populations, which are highly genetically divergent, giving the greatest room for variable drift. Observed changes in allele frequencies beyond extreme simulated values would then be more likely to contain an element of selective pressure.


```{r allSim99}
allSim99 <- allSim %>%
  lapply(quantile, probs = c(alpha/2, 1-alpha/2)) %>%
  lapply(as.list) %>%
  lapply(as.data.frame) %>%
  bind_rows() %>%
  set_names(c("lwr", "upr"))
```


```{r driftIntervals}
driftIntervals <- allParam %>%
  bind_rows() %>%
  dplyr::select(f0, n, sd) %>%
  as_tibble() %>%
  cbind(allSim99)
```

```{r driftBands, echo=FALSE, fig.cap=paste("Expected", percent(1-alpha), "intervals for individual allele frequencies under each set of criteria used in generation of simulations."), fig.height=6}
driftIntervals %>%
  mutate(
      n = as.factor(n),
      sd = as.factor(sd)
  ) %>%
    dplyr::select(f0, nPops = n, sd_0 = sd, lwr, upr) %>%
    ggplot(aes(f0, shape = nPops, colour = sd_0)) +
    geom_point(aes(y = lwr), size = 2) +
    geom_smooth(
        aes(y = lwr, linetype = nPops), 
        se = FALSE, method = "lm", formula = y~poly(x,2), 
        size = 0.5
    ) +
    geom_point(aes(y = upr), size = 2) +
    geom_smooth(
        aes(y = upr, linetype = nPops), 
        se = FALSE, method = "lm", formula = y~poly(x,2), 
        size = 0.5
    ) +
    scale_linetype_manual(values = c(1, 2)) +
    scale_colour_manual(values = c(rgb(0.1,0.1,0.8), rgb(0.9,0,0))) +
    labs(
        x = "Initial Value (f0)",
        y = paste(percent(1 -alpha), "Simulated Prediction Interval"),
        shape = "Number of Neighbouring Populations (n)",
        linetype = "Number of Neighbouring Populations (n)",
        colour = expression(
            paste("Initial Variability in Neighbouring Populations (", sigma, ")")
        )
    ) +
    theme(legend.position = "bottom")
```

## Comparison With Observed Data


<!-- Edited to here -->

```{r}
genotypeResults <- read_tsv("genotypeResults.tsv")
alleleResults <- read_tsv("alleleResults.tsv")
sigSnps <- c(
    filter(alleleResults, FDR < 0.05)$snpID,
    filter(genotypeResults, FDR < 0.05)$snpID
) %>%
    unique
bothSnps <- intersect(
    filter(alleleResults, FDR < 0.05)$snpID,
    filter(genotypeResults, FDR < 0.05)$snpID
)
```

```{r, echo=FALSE}
cp <- paste(
    "*Allele frequencies in the two populations.", 
    "SNPs considered as significant under either the Full Genotype or Allele Frequency models are highlighted.", 
    "SNPs considered significant under both models are labelled.",
    percent(1 - alpha), 
    "prediction intervals are indicated by the bands for differing simulation parameters.",
    "Intervals generated by 4 neighbouring populations are shown in red, whilst intervals generated by 8 neighbouring populations are shown in blue.",
    "Neighbouring populations simulated with the closest similarity to the main population are shown with solid lines (&sigma; = 0.5), whilst those with less similarity are shown as dashed lines (&sigma; = 0.8).",
    "The two points which clearly fall within all bands were found significant under the full genotype model and show differences in heterozygosity rather than in the allele frequencies themselves.",
    "The inner blue bands represent the expected scenario which most closely resembles the ecological system under investigation, with a continuous population of rabbits (i.e. a large number of neighbouring populations), with highly correlated allele frequencies.*"
    )
```


```{r exportFig4, echo=FALSE}
pdf(file.path("..", "figures", "Figure4.pdf"), width = w, height = 0.8*h)
alleleResults %>%
    mutate(
        Model = case_when(
            FDR < 0.05 & snpID %in% filter(genotypeResults, FDR < 0.05)$snpID ~ "Both Models", 
            FDR < 0.05 ~ "Allele Model",
            snpID %in% filter(genotypeResults, FDR < 0.05)$snpID ~ "Genotype Model",
            !snpID %in% filter(genotypeResults, FDR < 0.05)$snpID ~ "Not Significant"
        ),
        Model = as.factor(Model),
        Model = relevel(Model, ref = "Both Models")
    ) %>%
    ggplot() +
    geom_point(
        aes(x = MAF_1996, y = MAF_2012),
        data = . %>% filter(is.na(Model)),
        alpha = 0.4
    ) +
    geom_point(
        aes(x = MAF_1996, y = MAF_2012, colour = Model),
        data = . %>% filter(!is.na(Model))
    ) +
    geom_smooth(
        aes(x = f0, y = lwr, linetype = as.factor(sd)), 
        data = filter(driftIntervals, n == 4),
        method = "lm", formula = y ~ poly(x, 2),
        fullrange = TRUE, se = FALSE,
        colour = rgb(0.9,0,0), size = 0.8
    ) +
    geom_smooth(
        aes(x = f0, y = lwr, linetype = as.factor(sd)), 
        data = filter(driftIntervals, n == 8),
        method = "lm", formula = y ~ poly(x, 2),
        fullrange = TRUE, se = FALSE,
        colour = rgb(0.1,0.1,0.8), size = 0.8
    ) +
    geom_smooth(
        aes(x = f0, y = upr, linetype = as.factor(sd)), 
        data = filter(driftIntervals, n == 4),
        method = "lm", formula = y ~ poly(x, 2),
        fullrange = FALSE, se = FALSE,
        colour = rgb(0.9,0,0), size = 0.8
    ) +
    geom_smooth(
        aes(x = f0, y = upr, linetype = as.factor(sd)), 
        data = filter(driftIntervals, n == 8),
        method = "lm", formula = y ~ poly(x, 2),
        fullrange = FALSE, se = FALSE,
        colour = rgb(0.1,0.1,0.8), size = 0.8
    ) +
    geom_label_repel(
        aes(x = MAF_1996, y = MAF_2012, label = snpID, colour = Model),
        data = . %>% filter(Model == "Both Models"),
        size = 2,
        show.legend = FALSE
    ) +
    scale_colour_viridis_d(option = "B") +
    labs(
        x = expression(f[1996]),
        y = expression(f[2012]),
        colour = "Significance",
        linetype = expression(sigma)
    ) +
    # guides(linetype = FALSE) +
    theme(
        legend.position = "bottom",
        axis.text = element_text(size = 12),
        axis.title = element_text(size=  14)
    )
dev.off()
```

```{bash, echo = FALSE}
convert \
-density 600 \
-size 1920 \
../figures/Figure4.pdf \
../figures/Figure4.png
```

```{r flkWithDriftBands, echo=FALSE, fig.height=7, fig.cap=cp}
knitr::include_graphics(file.path("..", "figures", "Figure4.png"))
```


```{r}
save.image("06_GeneticDrift.RData")
```

# Session Information

```{r}
pander(sessionInfo())
```

